// @wordpress/components/build/range-control/input-range.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _rangeControlStyles = require("./styles/range-control-styles");

var _utils = require("./utils");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
const noop = () => {};

function InputRange(props, ref) {
  const {
    describedBy,
    label,
    onHideTooltip = noop,
    onMouseLeave = noop,
    onMouseMove = noop,
    onShowTooltip = noop,
    value,
    ...otherProps
  } = props;
  const hoverInteractions = (0, _utils.useDebouncedHoverInteraction)({
    onHide: onHideTooltip,
    onMouseLeave,
    onMouseMove,
    onShow: onShowTooltip
  });
  return (0, _element.createElement)(_rangeControlStyles.InputRange, (0, _extends2.default)({}, otherProps, hoverInteractions, {
    "aria-describedby": describedBy,
    "aria-label": label,
    "aria-hidden": false,
    ref: ref,
    tabIndex: 0,
    type: "range",
    value: value
  }));
}

const ForwardedComponent = (0, _element.forwardRef)(InputRange);
var _default = ForwardedComponent;
exports.default = _default;
