// @automattic/components/dist/cjs/segmented-control/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const clsx_1 = tslib_1.__importDefault(require("clsx"));
const prop_types_1 = tslib_1.__importDefault(require("prop-types"));
const react_1 = require("react");
const item_1 = tslib_1.__importDefault(require("./item"));
require("./style.scss");
class SegmentedControl extends react_1.Component {
    static Item = item_1.default;
    static propTypes = {
        className: prop_types_1.default.string,
        compact: prop_types_1.default.bool,
        primary: prop_types_1.default.bool,
        style: prop_types_1.default.object,
        children: prop_types_1.default.node.isRequired,
    };
    render() {
        const segmentedClasses = {
            'is-compact': this.props.compact,
            'is-primary': this.props.primary,
        };
        return ((0, jsx_runtime_1.jsx)("ul", { className: (0, clsx_1.default)('segmented-control', segmentedClasses, this.props.className), style: this.props.style, role: "radiogroup", children: this.props.children }));
    }
}
exports.default = SegmentedControl;
