// @woocommerce/components/build/calendar/phrases.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * External dependencies
 */
const i18n_1 = require("@wordpress/i18n");
exports.default = {
    calendarLabel: (0, i18n_1.__)('Calendar', 'woocommerce'),
    closeDatePicker: (0, i18n_1.__)('Close', 'woocommerce'),
    focusStartDate: (0, i18n_1.__)('Interact with the calendar and select start and end dates.', 'woocommerce'),
    clearDate: (0, i18n_1.__)('Clear Date', 'woocommerce'),
    clearDates: (0, i18n_1.__)('Clear Dates', 'woocommerce'),
    jumpToPrevMonth: (0, i18n_1.__)('Move backward to switch to the previous month.', 'woocommerce'),
    jumpToNextMonth: (0, i18n_1.__)('Move forward to switch to the next month.', 'woocommerce'),
    enterKey: (0, i18n_1.__)('Enter key', 'woocommerce'),
    leftArrowRightArrow: (0, i18n_1.__)('Right and left arrow keys', 'woocommerce'),
    upArrowDownArrow: (0, i18n_1.__)('up and down arrow keys', 'woocommerce'),
    pageUpPageDown: (0, i18n_1.__)('page up and page down keys', 'woocommerce'),
    homeEnd: (0, i18n_1.__)('Home and end keys', 'woocommerce'),
    escape: (0, i18n_1.__)('Escape key', 'woocommerce'),
    questionMark: (0, i18n_1.__)('Question mark', 'woocommerce'),
    selectFocusedDate: (0, i18n_1.__)('Select the date in focus.', 'woocommerce'),
    moveFocusByOneDay: (0, i18n_1.__)('Move backward (left) and forward (right) by one day.', 'woocommerce'),
    moveFocusByOneWeek: (0, i18n_1.__)('Move backward (up) and forward (down) by one week.', 'woocommerce'),
    moveFocusByOneMonth: (0, i18n_1.__)('Switch months.', 'woocommerce'),
    moveFocustoStartAndEndOfWeek: (0, i18n_1.__)('Go to the first or last day of a week.', 'woocommerce'),
    returnFocusToInput: (0, i18n_1.__)('Return to the date input field.', 'woocommerce'),
    keyboardNavigationInstructions: (0, i18n_1.__)('Press the down arrow key to interact with the calendar and select a date.', 'woocommerce'),
    chooseAvailableStartDate: ({ date }) => (0, i18n_1.sprintf)((0, i18n_1.__)('Select %s as a start date.', 'woocommerce'), date),
    chooseAvailableEndDate: ({ date }) => (0, i18n_1.sprintf)((0, i18n_1.__)('Select %s as an end date.', 'woocommerce'), date),
    chooseAvailableDate: ({ date }) => date,
    dateIsUnavailable: ({ date }) => (0, i18n_1.sprintf)((0, i18n_1.__)('%s is not selectable.', 'woocommerce'), date),
    dateIsSelected: ({ date }) => (0, i18n_1.sprintf)((0, i18n_1.__)('Selected. %s', 'woocommerce'), date),
};
