// @automattic/components/dist/cjs/button/index.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transparent = exports.Plain = exports.Link = exports.Disabled = exports.Borderless = exports.Scary = exports.Busy = exports.Compact = exports.Normal = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const addon_actions_1 = require("@storybook/addon-actions");
const _1 = tslib_1.__importDefault(require("."));
require("./stories.scss");
exports.default = { title: 'packages/components/Button' };
const helloWorld = `Hello World!`;
const handleClick = (0, addon_actions_1.action)('click');
const ButtonVariations = (props) => ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(_1.default, { onClick: handleClick, ...props, children: helloWorld }), ' ', (0, jsx_runtime_1.jsx)(_1.default, { primary: true, onClick: handleClick, ...props, children: helloWorld })] }));
const Normal = () => (0, jsx_runtime_1.jsx)(ButtonVariations, {});
exports.Normal = Normal;
const Compact = () => (0, jsx_runtime_1.jsx)(ButtonVariations, { compact: true });
exports.Compact = Compact;
const Busy = () => (0, jsx_runtime_1.jsx)(ButtonVariations, { busy: true });
exports.Busy = Busy;
const Scary = () => (0, jsx_runtime_1.jsx)(ButtonVariations, { scary: true });
exports.Scary = Scary;
const Borderless = () => (0, jsx_runtime_1.jsx)(ButtonVariations, { borderless: true });
exports.Borderless = Borderless;
const Disabled = () => (0, jsx_runtime_1.jsx)(ButtonVariations, { disabled: true });
exports.Disabled = Disabled;
const Link = () => (0, jsx_runtime_1.jsx)(ButtonVariations, { href: "https://www.google.com/", target: "_blank" });
exports.Link = Link;
const Plain = () => (0, jsx_runtime_1.jsx)(ButtonVariations, { plain: true });
exports.Plain = Plain;
const Transparent = () => {
    const wrapperStyle = { padding: '3em' };
    const sidebarStyle = {
        ...wrapperStyle,
        gap: '1em',
        backgroundColor: 'var( --color-sidebar-background )',
    };
    return ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)("div", { style: wrapperStyle, children: (0, jsx_runtime_1.jsx)(_1.default, { transparent: true, onClick: handleClick, children: helloWorld }) }), (0, jsx_runtime_1.jsx)("div", { className: "theme-default", style: sidebarStyle, children: (0, jsx_runtime_1.jsx)(_1.default, { transparent: true, onClick: handleClick, children: helloWorld }) })] }));
};
exports.Transparent = Transparent;
