// @automattic/calypso-analytics/dist/esm/utils/current-user.js
import hashPii from './hash-pii';
/**
 * Module variables
 */
let _currentUser;
/**
 * Gets current user.
 * @returns Current user.
 */
export function getCurrentUser() {
    return _currentUser;
}
/**
 * Sets current user, (stored in javascript memory).
 * @param currentUser the user data for the current user
 * @returns Current user.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function setCurrentUser(currentUser) {
    if (!currentUser.ID ||
        isNaN(parseInt(currentUser.ID, 10)) ||
        !currentUser.username ||
        !currentUser.email) {
        return; // Invalid user data.
    }
    _currentUser = {
        ID: parseInt(currentUser.ID, 10),
        username: currentUser.username,
        email: currentUser.email,
        localeSlug: currentUser.localeSlug,
        hashedPii: {
            ID: hashPii(currentUser.ID),
            username: hashPii(currentUser.username.toLowerCase().replace(/\s/g, '')),
            email: hashPii(currentUser.email.toLowerCase().replace(/\s/g, '')),
        },
    };
    return _currentUser;
}
