// @automattic/components/dist/cjs/jetpack-upsell-card/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const i18n_calypso_1 = require("i18n-calypso");
const react_1 = require("react");
const __1 = require("..");
const jetpack_icon_lock_svg_1 = tslib_1.__importDefault(require("../assets/jetpack-icon-lock.svg"));
const jetpack_product_icon_backup_svg_1 = tslib_1.__importDefault(require("../assets/jetpack-product-icon-backup.svg"));
const jetpack_product_icon_boost_svg_1 = tslib_1.__importDefault(require("../assets/jetpack-product-icon-boost.svg"));
const jetpack_product_icon_search_svg_1 = tslib_1.__importDefault(require("../assets/jetpack-product-icon-search.svg"));
const jetpack_product_icon_social_svg_1 = tslib_1.__importDefault(require("../assets/jetpack-product-icon-social.svg"));
const jetpack_product_icon_videopress_svg_1 = tslib_1.__importDefault(require("../assets/jetpack-product-icon-videopress.svg"));
require("./style.scss");
function JetpackUpsellCard({ purchasedProducts, siteSlug, upgradeUrls = {}, }) {
    // TODO: Make card collapsible
    const translate = (0, i18n_calypso_1.useTranslate)();
    const PRODUCTS = (0, react_1.useMemo)(() => [
        {
            description: translate('Protect your site from hackers and spam with automated backups, malware scanning, and spam filtering.'),
            href: 'https://jetpack.com/features/security/',
            iconUrl: jetpack_icon_lock_svg_1.default,
            isFree: false,
            slug: 'security',
            title: translate('Security', { context: 'Jetpack product name' }),
        },
        {
            description: translate('Save every single change and get back online quickly with one-click restores.'),
            href: 'https://jetpack.com/upgrade/backup/',
            iconUrl: jetpack_product_icon_backup_svg_1.default,
            isFree: false,
            slug: 'backup',
            title: translate('Backup'),
        },
        {
            description: translate("Help your site visitors instantly find what they're looking for so they read and buy more."),
            href: 'https://jetpack.com/upgrade/search/',
            iconUrl: jetpack_product_icon_search_svg_1.default,
            isFree: false,
            slug: 'search',
            title: translate('Search'),
        },
        {
            description: translate('Engage your visitors with high-quality, ad-free videos build specifically for WordPress.'),
            href: 'https://jetpack.com/videopress/',
            iconUrl: jetpack_product_icon_videopress_svg_1.default,
            isFree: false,
            slug: 'video',
            title: translate('VideoPress'),
        },
        {
            description: translate("Improve your site's performance and SEO in a few clicks with the free Jetpack Boost plugin."),
            href: 'https://jetpack.com/boost/',
            iconUrl: jetpack_product_icon_boost_svg_1.default,
            isFree: true,
            slug: 'boost',
            title: translate('Boost'),
        },
        {
            description: translate('Save time by auto-posting your content to social networks like Facebook, LinkedIn, and more.'),
            href: 'https://jetpack.com/social/',
            iconUrl: jetpack_product_icon_social_svg_1.default,
            isFree: true,
            slug: 'social',
            title: translate('Social'),
        },
        // TODO: Add Jetpack CRM upsell.
    ], [translate]);
    const visibleProducts = PRODUCTS.filter(({ slug }) => !purchasedProducts?.includes(slug) && slug in upgradeUrls);
    const hasProductsToUpsell = visibleProducts.length > 0;
    return !hasProductsToUpsell ? null : ((0, jsx_runtime_1.jsxs)(__1.Card, { className: "jetpack-upsell-card", children: [(0, jsx_runtime_1.jsxs)("h2", { className: "jetpack-upsell-card__title", children: [(0, jsx_runtime_1.jsx)("span", { className: "jetpack-upsell-card__title--long", children: translate('Enhance %(siteSlug)s with Jetpack Security, Performance, and Growth tools', {
                            args: { siteSlug: siteSlug ?? 'your site' },
                        }) }), (0, jsx_runtime_1.jsx)("span", { className: "jetpack-upsell-card__title--short", children: translate('Explore more tools by Jetpack') })] }), (0, jsx_runtime_1.jsx)("div", { className: "jetpack-upsell-card__content", children: visibleProducts.map(({ title, description, href, iconUrl, slug }) => ((0, jsx_runtime_1.jsxs)("div", { className: "jetpack-upsell-card__product", children: [(0, jsx_runtime_1.jsx)("div", { className: "jetpack-upsell-card__product-icon", children: (0, jsx_runtime_1.jsx)("img", { src: iconUrl, alt: title, width: "24px", height: "24px" }) }), (0, jsx_runtime_1.jsx)("h3", { className: "jetpack-upsell-card__product-title", children: title }), (0, jsx_runtime_1.jsx)("p", { className: "jetpack-upsell-card__product-description", children: description }), (0, jsx_runtime_1.jsxs)("a", { href: href, className: "jetpack-upsell-card__product-link", children: [(0, jsx_runtime_1.jsx)("span", { className: "jetpack-upsell-card__product-link-text", children: translate('More about %(productName)s', {
                                        args: { productName: title },
                                    }) }), (0, jsx_runtime_1.jsx)(__1.Gridicon, { icon: "external", size: 16 })] }), (0, jsx_runtime_1.jsx)(__1.Button, { href: upgradeUrls[slug], className: "jetpack-upsell-card__product-button", "aria-label": translate('Upgrade to Jetpack %(productName)s', {
                                args: { productName: title },
                            }), children: translate('Upgrade') })] }, slug))) })] }));
}
exports.default = JetpackUpsellCard;
