// @automattic/components/dist/cjs/select-dropdown/index.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Default = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const _1 = tslib_1.__importDefault(require("."));
exports.default = { component: _1.default, title: 'packages/components/SelectDropdown' };
const Template = (args) => {
    return (0, jsx_runtime_1.jsx)(_1.default, { ...args });
};
exports.Default = Template.bind({});
exports.Default.args = {
    ariaLabel: 'select-dropdown-aria-label',
    className: 'select-dropdown',
    compact: false,
    disabled: false,
    isLoading: false,
    // eslint-disable-next-line no-console
    onToggle: () => console.log('Toggled'),
    onSelect: ({ label }) => {
        // eslint-disable-next-line no-console
        console.log(`${label} Dropdown item selected`);
    },
    initialSelected: 'pikachu',
    options: [
        { value: 'pikachu', label: 'Pikachu' },
        { value: 'charmander', label: 'Charmander' },
        { value: 'bulbasaur', label: 'Bulbasaur' },
    ],
};
