// @automattic/components/dist/esm/upsell-menu-group/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import styled from '@emotion/styled';
import Gridicon from '../gridicon';
const Row = styled.div({
    display: 'flex',
    flexDirection: 'row',
    paddingTop: 8,
    paddingBottom: 8,
    gap: 8,
});
const Content = styled.div({
    display: 'flex',
    flexDirection: 'column',
    gap: 8,
    width: 210,
    whiteSpace: 'break-spaces',
    maxWidth: '80vw',
});
const BackgroundIcon = styled.div({
    backgroundColor: 'var(--color-accent)',
    width: 16,
    height: 16,
    borderRadius: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: 1,
});
const StarIcon = styled(Gridicon)({
    color: '#FFF',
});
export function UpsellIcon() {
    return (_jsx(BackgroundIcon, { children: _jsx(StarIcon, { icon: "star", size: 10 }) }));
}
export function UpsellMenuGroup(props) {
    const { icon = _jsx(UpsellIcon, {}), children, ...rest } = props;
    return (_jsxs(Row, { ...rest, children: [icon, _jsx(Content, { children: children })] }));
}
