// @woocommerce/components/build-module/abbreviated-card/index.js
/**
 * External dependencies
 */
import { createElement } from '@wordpress/element';
import { Card, CardBody, Icon } from '@wordpress/components';
import classnames from 'classnames';
import PropTypes from 'prop-types';
/**
 * Internal dependencies
 */
import Link from '../link';
const AbbreviatedCard = ({ children, className, href, icon, onClick, type, }) => {
    return (createElement(Card, { className: classnames('woocommerce-abbreviated-card', className) },
        createElement(CardBody, { size: null },
            createElement(Link, { href: href, onClick: onClick, type: type },
                createElement("div", { className: "woocommerce-abbreviated-card__icon" },
                    createElement(Icon, { icon: icon })),
                createElement("div", { className: "woocommerce-abbreviated-card__content" }, children)))));
};
AbbreviatedCard.propTypes = {
    /**
     * The Abbreviated Card content.
     */
    children: PropTypes.node.isRequired,
    /**
     * Additional CSS classes.
     */
    className: PropTypes.string,
    /**
     * The resource to link to.
     */
    href: PropTypes.string.isRequired,
    /**
     * Icon for the Abbreviated Card.
     */
    icon: PropTypes.element.isRequired,
    /**
     * Called when the card is clicked.
     */
    onClick: PropTypes.func,
    /**
     * Type of link.
     */
    type: PropTypes.oneOf(['wp-admin', 'wc-admin', 'external']),
};
export default AbbreviatedCard;
