// @automattic/components/dist/cjs/highlight-cards/stories/annual-highlight-cards.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnualHighlightCards = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const annual_highlight_cards_1 = tslib_1.__importDefault(require("../annual-highlight-cards"));
exports.default = {
    title: 'packages/components/Highlight Cards/AnnualHighlightCards',
    component: annual_highlight_cards_1.default,
    argTypes: {
        year: { control: 'number' },
        'counts.comments': { control: 'number' },
        'counts.likes': { control: 'number' },
        'counts.posts': { control: 'number' },
        'counts.words': { control: 'number' },
        'counts.followers': { control: 'number' },
    },
};
const Template = ({ year, ...counts }) => {
    const countsObject = {
        comments: counts['counts.comments'],
        likes: counts['counts.likes'],
        posts: counts['counts.posts'],
        words: counts['counts.words'],
        followers: counts['counts.followers'],
    };
    return ((0, jsx_runtime_1.jsx)("div", { style: { position: 'fixed', top: '50%', left: '50%', transform: 'translate(-50%, -50%)' }, children: (0, jsx_runtime_1.jsx)(annual_highlight_cards_1.default, { counts: countsObject, year: year }) }));
};
exports.AnnualHighlightCards = Template.bind({});
exports.AnnualHighlightCards.args = {
    year: 2022,
    'counts.comments': 72490,
    'counts.likes': 12298,
    'counts.posts': 79,
    'counts.words': 205035,
    'counts.followers': 1113323,
};
