// @automattic/components/dist/cjs/highlight-cards/stories/base-card.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCard_ = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const base_card_1 = tslib_1.__importDefault(require("../base-card"));
exports.default = {
    title: 'packages/components/Highlight Cards/BaseCard',
    component: base_card_1.default,
    argTypes: {
        heading: { control: 'text' },
        body: { control: 'text' },
    },
};
const Template = ({ heading, body }) => {
    return ((0, jsx_runtime_1.jsx)("div", { style: { position: 'fixed', top: '50%', left: '50%', transform: 'translate(-50%, -50%)' }, children: (0, jsx_runtime_1.jsx)(base_card_1.default, { heading: (0, jsx_runtime_1.jsx)("div", { children: heading }), children: (0, jsx_runtime_1.jsx)("div", { children: body }) }) }));
};
exports.BaseCard_ = Template.bind({});
exports.BaseCard_.args = {
    heading: 'Customizable Heading',
    body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
};
