// @automattic/components/dist/cjs/highlight-cards/stories/count-card.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CountCard_ = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_1 = require("@wordpress/icons");
const count_card_1 = tslib_1.__importDefault(require("../count-card"));
exports.default = {
    title: 'packages/components/Highlight Cards/CountCard',
    component: count_card_1.default,
    argTypes: {
        heading: { control: 'text' },
        valueType: {
            control: { type: 'radio' },
            options: ['number', 'string'],
        },
        stringValue: { control: 'text' },
        numberValue: { control: 'number' },
    },
};
const Template = (args) => {
    const value = args.valueType === 'string' ? args.stringValue : args.numberValue;
    return ((0, jsx_runtime_1.jsx)("div", { style: { position: 'fixed', top: '50%', left: '50%', transform: 'translate(-50%, -50%)' }, children: (0, jsx_runtime_1.jsx)(count_card_1.default, { heading: args.heading, icon: (0, jsx_runtime_1.jsx)(icons_1.Icon, { icon: icons_1.institution }), value: value }) }));
};
exports.CountCard_ = Template.bind({});
exports.CountCard_.args = {
    heading: 'Customizable Heading',
    valueType: 'number',
    stringValue: 'A really long string message that forces the box to expand',
    numberValue: 12345,
};
