// @automattic/components/dist/cjs/highlight-cards/stories/count-comparison-card.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CountComparisonCard_ = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_1 = require("@wordpress/icons");
const count_comparison_card_1 = tslib_1.__importDefault(require("../count-comparison-card"));
exports.default = {
    title: 'packages/components/Highlight Cards/CountComparisonCard',
    component: count_comparison_card_1.default,
    argTypes: {
        heading: { control: 'text' },
        previousCount: { control: 'number' },
        count: { control: 'number' },
        showValueTooltip: { control: 'boolean' },
    },
};
const Template = ({ count, previousCount, heading }) => {
    return ((0, jsx_runtime_1.jsx)("div", { style: { position: 'fixed', top: '50%', left: '50%', transform: 'translate(-50%, -50%)' }, children: (0, jsx_runtime_1.jsx)(count_comparison_card_1.default, { heading: heading, icon: (0, jsx_runtime_1.jsx)(icons_1.Icon, { icon: icons_1.institution }), count: count, previousCount: previousCount }) }));
};
exports.CountComparisonCard_ = Template.bind({});
exports.CountComparisonCard_.args = {
    heading: 'Customizable Heading',
    count: 234567,
    previousCount: 123456,
    showValueTooltip: false,
};
