// @automattic/components/dist/cjs/highlight-cards/stories/weekly-highlight.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeeklyHighlightCards = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const addon_actions_1 = require("@storybook/addon-actions");
const weekly_highlight_cards_1 = tslib_1.__importDefault(require("../weekly-highlight-cards"));
exports.default = {
    title: 'packages/components/Highlight Cards/WeeklyHighlightCards',
    component: weekly_highlight_cards_1.default,
    argTypes: {
        'counts.comments': { control: 'number' },
        'counts.likes': { control: 'number' },
        'counts.views': { control: 'number' },
        'counts.visitors': { control: 'number' },
        'previousCounts.comments': { control: 'number' },
        'previousCounts.likes': { control: 'number' },
        'previousCounts.views': { control: 'number' },
        'previousCounts.visitors': { control: 'number' },
        'story.isLoading': { control: 'boolean' },
        showValueTooltip: { control: 'boolean' },
    },
};
const handleClick = (0, addon_actions_1.action)('click');
const Template = ({ showValueTooltip, ...args }) => {
    const counts = args['story.isLoading']
        ? {}
        : {
            comments: args['counts.comments'],
            likes: args['counts.likes'],
            views: args['counts.views'],
            visitors: args['counts.visitors'],
        };
    const previousCounts = args['story.isLoading']
        ? {}
        : {
            comments: args['previousCounts.comments'],
            likes: args['previousCounts.likes'],
            views: args['previousCounts.views'],
            visitors: args['previousCounts.visitors'],
        };
    return ((0, jsx_runtime_1.jsx)("div", { style: { position: 'fixed', top: '50%', left: '50%', transform: 'translate(-50%, -50%)' }, children: (0, jsx_runtime_1.jsx)(weekly_highlight_cards_1.default, { counts: counts, previousCounts: previousCounts, onClickComments: handleClick, onClickLikes: handleClick, onClickViews: handleClick, onClickVisitors: handleClick, showValueTooltip: showValueTooltip }) }));
};
exports.WeeklyHighlightCards = Template.bind({});
exports.WeeklyHighlightCards.args = {
    'counts.comments': 45,
    'counts.likes': 0,
    'counts.views': 4673,
    'counts.visitors': 1548,
    'previousCounts.comments': 45,
    'previousCounts.likes': 100,
    'previousCounts.views': 4073,
    'previousCounts.visitors': 1412,
    showValueTooltip: true,
    'story.isLoading': false,
};
