// @automattic/i18n-utils/dist/cjs/locales.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jetpackComLocales = exports.magnificentNonEnLocales = exports.forumLocales = exports.supportSiteLocales = exports.livechatSupportLocales = exports.englishLocales = exports.localesToSubdomains = exports.localesForPricePlans = exports.localesWithLearn = exports.localesWithCookiePolicy = exports.localesWithPrivacyPolicy = exports.localesWithGoBlog = exports.localesWithBlog = exports.i18nDefaultLocaleSlug = void 0;
exports.i18nDefaultLocaleSlug = 'en';
exports.localesWithBlog = ['en', 'ja', 'es', 'pt', 'fr', 'pt-br'];
exports.localesWithGoBlog = ['en', 'pt-br', 'de', 'es', 'fr', 'it'];
exports.localesWithPrivacyPolicy = ['en', 'fr', 'de', 'es'];
exports.localesWithCookiePolicy = ['en', 'fr', 'de', 'es'];
exports.localesWithLearn = ['en', 'es'];
exports.localesForPricePlans = [
    'ar',
    'de',
    'el',
    'es',
    'fr',
    'he',
    'id',
    'it',
    'ja',
    'ko',
    'nl',
    'pt-br',
    'ro',
    'ru',
    'sv',
    'tr',
    'zh-cn',
    'zh-tw',
];
exports.localesToSubdomains = {
    'pt-br': 'br',
    br: 'bre',
    zh: 'zh-cn',
    'zh-hk': 'zh-tw',
    'zh-sg': 'zh-cn',
    kr: 'ko',
};
// replaces config( 'english_locales' )
exports.englishLocales = ['en', 'en-gb'];
// replaces config( 'livechat_support_locales' )
exports.livechatSupportLocales = ['en'];
// replaces config( 'support_site_locales' )
exports.supportSiteLocales = [
    'ar',
    'de',
    'en',
    'es',
    'fr',
    'he',
    'id',
    'it',
    'ja',
    'ko',
    'nl',
    'pt-br',
    'ru',
    'sv',
    'tr',
    'zh-cn',
    'zh-tw',
];
// replaces config( 'forum_locales')
exports.forumLocales = [
    'ar',
    'de',
    'el',
    'en',
    'es',
    'fa',
    'fi',
    'fr',
    'id',
    'it',
    'ja',
    'nl',
    'pt',
    'pt-br',
    'ru',
    'sv',
    'th',
    'tl',
    'tr',
];
// replaces config( 'magnificent_non_en_locales')
exports.magnificentNonEnLocales = [
    'es',
    'pt-br',
    'de',
    'fr',
    'he',
    'ja',
    'it',
    'nl',
    'ru',
    'tr',
    'id',
    'zh-cn',
    'zh-tw',
    'ko',
    'ar',
    'sv',
];
// replaces config( 'jetpack_com_locales')
exports.jetpackComLocales = [
    'en',
    'ar',
    'de',
    'es',
    'fr',
    'he',
    'id',
    'it',
    'ja',
    'ko',
    'nl',
    'pt-br',
    'ro',
    'ru',
    'sv',
    'tr',
    'zh-cn',
    'zh-tw',
];
