// @automattic/components/dist/esm/button/index.stories.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { action } from '@storybook/addon-actions';
import Button from '.';
import './stories.scss';
export default { title: 'packages/components/Button' };
const helloWorld = `Hello World!`;
const handleClick = action('click');
const ButtonVariations = (props) => (_jsxs(_Fragment, { children: [_jsx(Button, { onClick: handleClick, ...props, children: helloWorld }), ' ', _jsx(Button, { primary: true, onClick: handleClick, ...props, children: helloWorld })] }));
export const Normal = () => _jsx(ButtonVariations, {});
export const Compact = () => _jsx(ButtonVariations, { compact: true });
export const Busy = () => _jsx(ButtonVariations, { busy: true });
export const Scary = () => _jsx(ButtonVariations, { scary: true });
export const Borderless = () => _jsx(ButtonVariations, { borderless: true });
export const Disabled = () => _jsx(ButtonVariations, { disabled: true });
export const Link = () => _jsx(ButtonVariations, { href: "https://www.google.com/", target: "_blank" });
export const Plain = () => _jsx(ButtonVariations, { plain: true });
export const Transparent = () => {
    const wrapperStyle = { padding: '3em' };
    const sidebarStyle = {
        ...wrapperStyle,
        gap: '1em',
        backgroundColor: 'var( --color-sidebar-background )',
    };
    return (_jsxs(_Fragment, { children: [_jsx("div", { style: wrapperStyle, children: _jsx(Button, { transparent: true, onClick: handleClick, children: helloWorld }) }), _jsx("div", { className: "theme-default", style: sidebarStyle, children: _jsx(Button, { transparent: true, onClick: handleClick, children: helloWorld }) })] }));
};
