// @woocommerce/components/build/advanced-filters/number-filter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * External dependencies
 */
const element_1 = require("@wordpress/element");
const components_1 = require("@wordpress/components");
const lodash_1 = require("lodash");
const interpolate_components_1 = __importDefault(require("@automattic/interpolate-components"));
const classnames_1 = __importDefault(require("classnames"));
const i18n_1 = require("@wordpress/i18n");
const currency_1 = __importDefault(require("@woocommerce/currency"));
/**
 * Internal dependencies
 */
const text_control_with_affixes_1 = __importDefault(require("../text-control-with-affixes"));
const utils_1 = require("./utils");
class NumberFilter extends element_1.Component {
    getBetweenString() {
        return (0, i18n_1._x)('{{rangeStart /}}{{span}} and {{/span}}{{rangeEnd /}}', 'Numerical range inputs arranged on a single line', 'woocommerce');
    }
    getScreenReaderText(filter, config) {
        const { currency } = this.props;
        const rule = (0, lodash_1.find)(config.rules, { value: filter.rule }) || {};
        let [rangeStart, rangeEnd] = (0, lodash_1.isArray)(filter.value)
            ? filter.value
            : [filter.value];
        // Return nothing if we're missing input(s)
        if (!rangeStart || (rule.value === 'between' && !rangeEnd)) {
            return '';
        }
        const inputType = (0, lodash_1.get)(config, ['input', 'type'], 'number');
        if (inputType === 'currency') {
            const { formatAmount } = (0, currency_1.default)(currency);
            rangeStart = formatAmount(rangeStart);
            rangeEnd = formatAmount(rangeEnd);
        }
        let filterStr = rangeStart;
        if (rule.value === 'between') {
            filterStr = (0, interpolate_components_1.default)({
                mixedString: this.getBetweenString(),
                components: {
                    rangeStart: (0, element_1.createElement)(element_1.Fragment, null, rangeStart),
                    rangeEnd: (0, element_1.createElement)(element_1.Fragment, null, rangeEnd),
                    span: (0, element_1.createElement)(element_1.Fragment, null),
                },
            });
        }
        return (0, utils_1.textContent)((0, interpolate_components_1.default)({
            mixedString: config.labels.title,
            components: {
                filter: (0, element_1.createElement)(element_1.Fragment, null, filterStr),
                rule: (0, element_1.createElement)(element_1.Fragment, null, rule.label),
                title: (0, element_1.createElement)(element_1.Fragment, null),
            },
        }));
    }
    getFormControl({ type, value, label, onChange, currencySymbol, symbolPosition, }) {
        if (type === 'currency') {
            return symbolPosition.indexOf('right') === 0 ? ((0, element_1.createElement)(text_control_with_affixes_1.default, { suffix: (0, element_1.createElement)("span", { dangerouslySetInnerHTML: {
                        __html: currencySymbol,
                    } }), className: "woocommerce-filters-advanced__input", type: "number", value: value || '', "aria-label": label, onChange: onChange })) : ((0, element_1.createElement)(text_control_with_affixes_1.default, { prefix: (0, element_1.createElement)("span", { dangerouslySetInnerHTML: {
                        __html: currencySymbol,
                    } }), className: "woocommerce-filters-advanced__input", type: "number", value: value || '', "aria-label": label, onChange: onChange }));
        }
        return ((0, element_1.createElement)(components_1.TextControl, { className: "woocommerce-filters-advanced__input", type: "number", value: value || '', "aria-label": label, onChange: onChange }));
    }
    getFilterInputs() {
        const { config, filter, onFilterChange, currency } = this.props;
        const { symbol: currencySymbol, symbolPosition } = currency;
        if (filter.rule === 'between') {
            return this.getRangeInput();
        }
        const inputType = (0, lodash_1.get)(config, ['input', 'type'], 'number');
        const [rangeStart, rangeEnd] = (0, lodash_1.isArray)(filter.value)
            ? filter.value
            : [filter.value];
        if (Boolean(rangeEnd)) {
            // If there's a value for rangeEnd, we've just changed from "between"
            // to "less than" or "more than" and need to transition the value
            onFilterChange({
                property: 'value',
                value: rangeStart || rangeEnd,
            });
        }
        let labelFormat = '';
        if (filter.rule === 'lessthan') {
            /* eslint-disable-next-line max-len */
            /* translators: Sentence fragment, "maximum amount" refers to a numeric value the field must be less than. Screenshot for context: https://cloudup.com/cmv5CLyMPNQ */
            labelFormat = (0, i18n_1._x)('%(field)s maximum amount', 'maximum value input', 'woocommerce');
        }
        else {
            /* eslint-disable-next-line max-len */
            /* translators: Sentence fragment, "minimum amount" refers to a numeric value the field must be more than. Screenshot for context: https://cloudup.com/cmv5CLyMPNQ */
            labelFormat = (0, i18n_1._x)('%(field)s minimum amount', 'minimum value input', 'woocommerce');
        }
        return this.getFormControl({
            type: inputType,
            value: rangeStart || rangeEnd,
            label: (0, i18n_1.sprintf)(labelFormat, {
                field: (0, lodash_1.get)(config, ['labels', 'add']),
            }),
            onChange: (value) => onFilterChange({ property: 'value', value }),
            currencySymbol,
            symbolPosition,
        });
    }
    getRangeInput() {
        const { config, filter, onFilterChange, currency } = this.props;
        const { symbol: currencySymbol, symbolPosition } = currency;
        const inputType = (0, lodash_1.get)(config, ['input', 'type'], 'number');
        const [rangeStart, rangeEnd] = (0, lodash_1.isArray)(filter.value)
            ? filter.value
            : [filter.value];
        const rangeStartOnChange = (newRangeStart) => {
            onFilterChange({
                property: 'value',
                key: [newRangeStart, rangeEnd],
            });
        };
        const rangeEndOnChange = (newRangeEnd) => {
            onFilterChange({
                property: 'value',
                key: [rangeStart, newRangeEnd],
            });
        };
        return (0, interpolate_components_1.default)({
            mixedString: this.getBetweenString(),
            components: {
                rangeStart: this.getFormControl({
                    type: inputType,
                    value: rangeStart || '',
                    label: (0, i18n_1.sprintf)(
                    /* eslint-disable-next-line max-len */
                    /* translators: Sentence fragment, "range start" refers to the first of two numeric values the field must be between. Screenshot for context: https://cloudup.com/cmv5CLyMPNQ */
                    (0, i18n_1.__)('%(field)s range start', 'woocommerce'), { field: (0, lodash_1.get)(config, ['labels', 'add']) }),
                    onChange: rangeStartOnChange,
                    currencySymbol,
                    symbolPosition,
                }),
                rangeEnd: this.getFormControl({
                    type: inputType,
                    value: rangeEnd || '',
                    label: (0, i18n_1.sprintf)(
                    /* eslint-disable-next-line max-len */
                    /* translators: Sentence fragment, "range end" refers to the second of two numeric values the field must be between. Screenshot for context: https://cloudup.com/cmv5CLyMPNQ */
                    (0, i18n_1.__)('%(field)s range end', 'woocommerce'), { field: (0, lodash_1.get)(config, ['labels', 'add']) }),
                    onChange: rangeEndOnChange,
                    currencySymbol,
                    symbolPosition,
                }),
                span: (0, element_1.createElement)("span", { className: "separator" }),
            },
        });
    }
    render() {
        const { className, config, filter, onFilterChange, isEnglish } = this.props;
        const { rule } = filter;
        const { labels, rules } = config;
        const children = (0, interpolate_components_1.default)({
            mixedString: labels.title,
            components: {
                title: (0, element_1.createElement)("span", { className: className }),
                rule: ((0, element_1.createElement)(components_1.SelectControl, { className: (0, classnames_1.default)(className, 'woocommerce-filters-advanced__rule'), options: rules, value: rule, onChange: (value) => onFilterChange({ property: 'rule', value }), "aria-label": labels.rule })),
                filter: ((0, element_1.createElement)("div", { className: (0, classnames_1.default)(className, 'woocommerce-filters-advanced__input-range', {
                        'is-between': rule === 'between',
                    }) }, this.getFilterInputs())),
            },
        });
        const screenReaderText = this.getScreenReaderText(filter, config);
        /*eslint-disable jsx-a11y/no-noninteractive-tabindex*/
        return ((0, element_1.createElement)("fieldset", { className: "woocommerce-filters-advanced__line-item", tabIndex: "0" },
            (0, element_1.createElement)("legend", { className: "screen-reader-text" }, labels.add || ''),
            (0, element_1.createElement)("div", { className: (0, classnames_1.default)('woocommerce-filters-advanced__fieldset', {
                    'is-english': isEnglish,
                }) }, children),
            screenReaderText && ((0, element_1.createElement)("span", { className: "screen-reader-text" }, screenReaderText))));
        /*eslint-enable jsx-a11y/no-noninteractive-tabindex*/
    }
}
exports.default = NumberFilter;
