// @woocommerce/components/build/compare-filter/button.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * External dependencies
 */
const classnames_1 = __importDefault(require("classnames"));
const prop_types_1 = __importDefault(require("prop-types"));
const components_1 = require("@wordpress/components");
const element_1 = require("@wordpress/element");
/**
 * A button used when comparing items, if `count` is less than 2 a hoverable tooltip is added with `helpText`.
 *
 * @param {Object}   props
 * @param {string}   props.className
 * @param {number}   props.count
 * @param {Node}     props.children
 * @param {boolean}  props.disabled
 * @param {string}   props.helpText
 * @param {Function} props.onClick
 * @return {Object} -
 */
const CompareButton = ({ className, count, children, disabled, helpText, onClick, }) => !disabled && count < 2 ? ((0, element_1.createElement)(components_1.Tooltip, { text: helpText },
    (0, element_1.createElement)("span", { className: className },
        (0, element_1.createElement)(components_1.Button, { className: "woocommerce-compare-button", disabled: true, isSecondary: true }, children)))) : ((0, element_1.createElement)(components_1.Button, { className: (0, classnames_1.default)('woocommerce-compare-button', className), onClick: onClick, disabled: disabled, isSecondary: true }, children));
CompareButton.propTypes = {
    /**
     * Additional CSS classes.
     */
    className: prop_types_1.default.string,
    /**
     * The count of items selected.
     */
    count: prop_types_1.default.number.isRequired,
    /**
     * The button content.
     */
    children: prop_types_1.default.node.isRequired,
    /**
     * Text displayed when hovering over a disabled button.
     */
    helpText: prop_types_1.default.string.isRequired,
    /**
     * The function called when the button is clicked.
     */
    onClick: prop_types_1.default.func.isRequired,
    /**
     * Whether the control is disabled or not.
     */
    disabled: prop_types_1.default.bool,
};
exports.default = CompareButton;
