// @automattic/i18n-utils/dist/esm/locales.js
export const i18nDefaultLocaleSlug = 'en';
export const localesWithBlog = ['en', 'ja', 'es', 'pt', 'fr', 'pt-br'];
export const localesWithGoBlog = ['en', 'pt-br', 'de', 'es', 'fr', 'it'];
export const localesWithPrivacyPolicy = ['en', 'fr', 'de', 'es'];
export const localesWithCookiePolicy = ['en', 'fr', 'de', 'es'];
export const localesWithLearn = ['en', 'es'];
export const localesForPricePlans = [
    'ar',
    'de',
    'el',
    'es',
    'fr',
    'he',
    'id',
    'it',
    'ja',
    'ko',
    'nl',
    'pt-br',
    'ro',
    'ru',
    'sv',
    'tr',
    'zh-cn',
    'zh-tw',
];
export const localesToSubdomains = {
    'pt-br': 'br',
    br: 'bre',
    zh: 'zh-cn',
    'zh-hk': 'zh-tw',
    'zh-sg': 'zh-cn',
    kr: 'ko',
};
// replaces config( 'english_locales' )
export const englishLocales = ['en', 'en-gb'];
// replaces config( 'livechat_support_locales' )
export const livechatSupportLocales = ['en'];
// replaces config( 'support_site_locales' )
export const supportSiteLocales = [
    'ar',
    'de',
    'en',
    'es',
    'fr',
    'he',
    'id',
    'it',
    'ja',
    'ko',
    'nl',
    'pt-br',
    'ru',
    'sv',
    'tr',
    'zh-cn',
    'zh-tw',
];
// replaces config( 'forum_locales')
export const forumLocales = [
    'ar',
    'de',
    'el',
    'en',
    'es',
    'fa',
    'fi',
    'fr',
    'id',
    'it',
    'ja',
    'nl',
    'pt',
    'pt-br',
    'ru',
    'sv',
    'th',
    'tl',
    'tr',
];
// replaces config( 'magnificent_non_en_locales')
export const magnificentNonEnLocales = [
    'es',
    'pt-br',
    'de',
    'fr',
    'he',
    'ja',
    'it',
    'nl',
    'ru',
    'tr',
    'id',
    'zh-cn',
    'zh-tw',
    'ko',
    'ar',
    'sv',
];
// replaces config( 'jetpack_com_locales')
export const jetpackComLocales = [
    'en',
    'ar',
    'de',
    'es',
    'fr',
    'he',
    'id',
    'it',
    'ja',
    'ko',
    'nl',
    'pt-br',
    'ro',
    'ru',
    'sv',
    'tr',
    'zh-cn',
    'zh-tw',
];
