// @automattic/components/dist/cjs/notice-banner/notice-banner.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.error = exports.success = exports.warning = exports._default = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const components_1 = require("@wordpress/components");
const button_1 = tslib_1.__importDefault(require("../button"));
const index_1 = tslib_1.__importDefault(require("./index"));
exports.default = {
    title: 'packages/components/Notice Banner',
    component: index_1.default,
    argTypes: {
        level: {
            control: {
                type: 'select',
                options: ['info', 'success', 'warning', 'error'],
            },
        },
        hideCloseButton: {
            control: {
                type: 'boolean',
            },
        },
    },
};
const Template = (args) => (0, jsx_runtime_1.jsx)(index_1.default, { ...args });
exports._default = Template.bind({});
exports._default.args = {
    level: 'info',
    title: 'Improve your hovercraft',
    children: 'Make your hovercraft better with HoverPack; the best hovercraft upgrade kit on the market.',
    onClose: () => alert('Close clicked'),
    actions: [
        (0, jsx_runtime_1.jsx)(button_1.default, { primary: true, children: "Install now" }, "install"),
        (0, jsx_runtime_1.jsx)(components_1.ExternalLink, { href: "https://en.wikipedia.org/wiki/Hovercraft", children: "Learn more" }, "learn-more"),
    ],
    hideCloseButton: false,
};
exports.warning = Template.bind({});
exports.warning.args = {
    level: 'warning',
    title: 'Your hovercraft is full of eels.',
    children: ((0, jsx_runtime_1.jsx)("div", { children: "Either your vehicle needs to be cleared, or some kind of translation error has occurred." })),
    actions: [
        (0, jsx_runtime_1.jsx)(button_1.default, { primary: true, children: "Start Bailing" }, "bail"),
        (0, jsx_runtime_1.jsx)(components_1.ExternalLink, { href: "https://en.wikipedia.org/wiki/Hovercraft", children: "Learn more" }, "learn-more"),
    ],
    hideCloseButton: false,
};
exports.success = Template.bind({});
exports.success.args = {
    level: 'success',
    title: 'Your hovercraft has been upgraded.',
    children: 'Please enjoy your newer, cooler hovercraft.',
    onClose: () => alert('Close clicked'),
    actions: [
        (0, jsx_runtime_1.jsx)(button_1.default, { isPrimary: true, children: "Start crafting" }, "start"),
    ],
    hideCloseButton: false,
};
exports.error = Template.bind({});
exports.error.args = {
    level: 'error',
    title: 'The eels have stolen your hovercraft.',
    children: 'We were unable to remove the eels from your hovercraft. Please contact the authorities, as the eels are armed and dangerous.',
    onClose: () => alert('Close clicked'),
    actions: [
        (0, jsx_runtime_1.jsx)(components_1.ExternalLink, { href: "https://en.wikipedia.org/wiki/Eel", children: "Learn more" }, "learn-more"),
    ],
    hideCloseButton: false,
};
