// @automattic/components/dist/esm/jetpack-upsell-card/index.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useTranslate } from 'i18n-calypso';
import { useMemo } from 'react';
import { Button, Card, Gridicon } from '..';
import SecurityIcon from '../assets/jetpack-icon-lock.svg';
import BackupIcon from '../assets/jetpack-product-icon-backup.svg';
import BoostIcon from '../assets/jetpack-product-icon-boost.svg';
import SearchIcon from '../assets/jetpack-product-icon-search.svg';
import SocialIcon from '../assets/jetpack-product-icon-social.svg';
import VideoPressIcon from '../assets/jetpack-product-icon-videopress.svg';
import './style.scss';
export default function JetpackUpsellCard({ purchasedProducts, siteSlug, upgradeUrls = {}, }) {
    // TODO: Make card collapsible
    const translate = useTranslate();
    const PRODUCTS = useMemo(() => [
        {
            description: translate('Protect your site from hackers and spam with automated backups, malware scanning, and spam filtering.'),
            href: 'https://jetpack.com/features/security/',
            iconUrl: SecurityIcon,
            isFree: false,
            slug: 'security',
            title: translate('Security', { context: 'Jetpack product name' }),
        },
        {
            description: translate('Save every single change and get back online quickly with one-click restores.'),
            href: 'https://jetpack.com/upgrade/backup/',
            iconUrl: BackupIcon,
            isFree: false,
            slug: 'backup',
            title: translate('Backup'),
        },
        {
            description: translate("Help your site visitors instantly find what they're looking for so they read and buy more."),
            href: 'https://jetpack.com/upgrade/search/',
            iconUrl: SearchIcon,
            isFree: false,
            slug: 'search',
            title: translate('Search'),
        },
        {
            description: translate('Engage your visitors with high-quality, ad-free videos build specifically for WordPress.'),
            href: 'https://jetpack.com/videopress/',
            iconUrl: VideoPressIcon,
            isFree: false,
            slug: 'video',
            title: translate('VideoPress'),
        },
        {
            description: translate("Improve your site's performance and SEO in a few clicks with the free Jetpack Boost plugin."),
            href: 'https://jetpack.com/boost/',
            iconUrl: BoostIcon,
            isFree: true,
            slug: 'boost',
            title: translate('Boost'),
        },
        {
            description: translate('Save time by auto-posting your content to social networks like Facebook, LinkedIn, and more.'),
            href: 'https://jetpack.com/social/',
            iconUrl: SocialIcon,
            isFree: true,
            slug: 'social',
            title: translate('Social'),
        },
        // TODO: Add Jetpack CRM upsell.
    ], [translate]);
    const visibleProducts = PRODUCTS.filter(({ slug }) => !purchasedProducts?.includes(slug) && slug in upgradeUrls);
    const hasProductsToUpsell = visibleProducts.length > 0;
    return !hasProductsToUpsell ? null : (_jsxs(Card, { className: "jetpack-upsell-card", children: [_jsxs("h2", { className: "jetpack-upsell-card__title", children: [_jsx("span", { className: "jetpack-upsell-card__title--long", children: translate('Enhance %(siteSlug)s with Jetpack Security, Performance, and Growth tools', {
                            args: { siteSlug: siteSlug ?? 'your site' },
                        }) }), _jsx("span", { className: "jetpack-upsell-card__title--short", children: translate('Explore more tools by Jetpack') })] }), _jsx("div", { className: "jetpack-upsell-card__content", children: visibleProducts.map(({ title, description, href, iconUrl, slug }) => (_jsxs("div", { className: "jetpack-upsell-card__product", children: [_jsx("div", { className: "jetpack-upsell-card__product-icon", children: _jsx("img", { src: iconUrl, alt: title, width: "24px", height: "24px" }) }), _jsx("h3", { className: "jetpack-upsell-card__product-title", children: title }), _jsx("p", { className: "jetpack-upsell-card__product-description", children: description }), _jsxs("a", { href: href, className: "jetpack-upsell-card__product-link", children: [_jsx("span", { className: "jetpack-upsell-card__product-link-text", children: translate('More about %(productName)s', {
                                        args: { productName: title },
                                    }) }), _jsx(Gridicon, { icon: "external", size: 16 })] }), _jsx(Button, { href: upgradeUrls[slug], className: "jetpack-upsell-card__product-button", "aria-label": translate('Upgrade to Jetpack %(productName)s', {
                                args: { productName: title },
                            }), children: translate('Upgrade') })] }, slug))) })] }));
}
