// @automattic/calypso-analytics/dist/esm/train-tracks.js
import { v4 as uuid } from 'uuid';
import { recordTracksEvent } from './tracks';
export function recordTrainTracksRender({ railcarId, uiAlgo, uiPosition, fetchAlgo, fetchPosition, query, result, recBlogId, recPostId, recFeedId, recFeedItemId, }) {
    const props = {};
    // Remap and filter undefined props
    Object.entries({
        railcar: railcarId,
        ui_algo: uiAlgo,
        ui_position: uiPosition,
        fetch_algo: fetchAlgo,
        fetch_query: query,
        fetch_position: fetchPosition,
        rec_result: result,
        rec_blog_id: recBlogId,
        rec_post_id: recPostId,
        rec_feed_id: recFeedId,
        rec_feed_item_id: recFeedItemId,
    }).forEach(([key, val]) => val !== undefined && (props[key] = val));
    recordTracksEvent('calypso_traintracks_render', props);
}
export function recordTrainTracksInteract({ railcarId, action }) {
    recordTracksEvent('calypso_traintracks_interact', {
        railcar: railcarId,
        action,
    });
}
export function getNewRailcarId(suffix = 'recommendation') {
    return `${uuid().replace(/-/g, '')}-${suffix}`;
}
