// @automattic/components/dist/cjs/post-stats-card/stories/index.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithoutAThumbnail = exports.FullWidth = exports.Default = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const __1 = tslib_1.__importDefault(require("../"));
exports.default = { title: 'packages/components/Post Stats Card' };
const PostStatsCardVariations = (props) => ((0, jsx_runtime_1.jsx)(__1.default, { commentCount: 7, heading: "Latest post", likeCount: 83, viewCount: 912, post: {
        date: '2022-12-07T14:24:48+00:00',
        post_thumbnail: 'https://images.unsplash.com/photo-1670510521067-2e94183f9803',
        title: 'The moody trail',
    }, ...props }));
const Default = () => ((0, jsx_runtime_1.jsx)("div", { className: "post-stats-card-story", style: { margin: 'auto', maxWidth: '700px' }, children: (0, jsx_runtime_1.jsx)(PostStatsCardVariations, {}) }));
exports.Default = Default;
const FullWidth = () => ((0, jsx_runtime_1.jsx)("div", { className: "post-stats-card-story", children: (0, jsx_runtime_1.jsx)(PostStatsCardVariations, {}) }));
exports.FullWidth = FullWidth;
const WithoutAThumbnail = () => ((0, jsx_runtime_1.jsx)("div", { className: "post-stats-card-story", children: (0, jsx_runtime_1.jsx)(PostStatsCardVariations, { post: {
            date: '2022-12-07T14:24:48+00:00',
            post_thumbnail: null,
            title: 'The moody trail',
        } }) }));
exports.WithoutAThumbnail = WithoutAThumbnail;
