// @automattic/components/dist/cjs/chart-uplot/bar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const components_1 = require("@wordpress/components");
const react_1 = require("react");
const uplot_react_1 = tslib_1.__importDefault(require("uplot-react"));
const popover_1 = tslib_1.__importDefault(require("../popover"));
const use_resize_1 = tslib_1.__importDefault(require("./hooks/use-resize"));
const multi_bars_1 = tslib_1.__importDefault(require("./uplot-plugins/multi-bars"));
// NOTE: Do not include this component in the package entry bundle.
//       Doing so will cause tests of the consumer package to break due to uPlot's reliance on matchMedia.
//       https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom.
const DEFAULT_DIMENSIONS = {
    height: 600,
    width: 1224,
};
function UplotBarChart({ data, labels, legendData = [], options: propOptions, isLoading = false, }) {
    const uplot = (0, react_1.useRef)(null);
    const uplotContainer = (0, react_1.useRef)(null);
    const [chartDimensions, setChartDimensions] = (0, react_1.useState)(DEFAULT_DIMENSIONS);
    const [legendVisibleIndex, setLegendVisibleIndex] = (0, react_1.useState)(-1);
    const [legendEl, setLegendEl] = (0, react_1.useState)(null);
    const options = (0, react_1.useMemo)(() => {
        const defaultOptions = {
            class: 'calypso-uplot-bar-chart',
            ...chartDimensions,
            axes: [
                {},
                {
                    show: true,
                    side: 1,
                },
            ],
            padding: [null, 0, null, 0],
            series: data[0].map((label, i) => ({
                label,
                fill: legendData[i].fillColor,
            })),
            plugins: [
                (0, multi_bars_1.default)({
                    labels: () => labels,
                }),
            ],
            legend: {
                live: false,
                isolate: true,
                mount: (self, el) => {
                    setLegendEl(el);
                },
            },
        };
        return {
            ...defaultOptions,
            ...(typeof propOptions === 'object' ? propOptions : {}),
        };
    }, [chartDimensions, data, labels, legendData, propOptions]);
    (0, use_resize_1.default)(uplot, uplotContainer);
    (0, react_1.useEffect)(() => {
        // Need extra check for container resize due to `seriesBarsPlugin`
        if (uplotContainer.current) {
            const { width, height } = uplotContainer.current.getBoundingClientRect();
            if (width !== chartDimensions.width || height !== chartDimensions.height) {
                setChartDimensions({ width, height });
            }
        }
    }, [chartDimensions, data]);
    (0, react_1.useEffect)(() => {
        if (legendEl === null) {
            return;
        }
        function onMouseOverListener(event) {
            const target = event.target;
            const legendDirectChildren = target?.closest('.u-series');
            if (!target || !legendDirectChildren || !legendEl) {
                return;
            }
            if (target.classList.contains('u-label')) {
                const seriesIdx = Array.from(legendEl.children).indexOf(legendDirectChildren);
                setLegendVisibleIndex(seriesIdx);
            }
        }
        function onMouseLeaveListener(event) {
            const target = event.target;
            if (!target) {
                return;
            }
            if (target.classList.contains('u-label')) {
                setLegendVisibleIndex(-1);
            }
        }
        legendEl.addEventListener('mouseover', onMouseOverListener);
        legendEl.addEventListener('mouseout', onMouseLeaveListener);
        return () => {
            legendEl.removeEventListener('mouseover', onMouseOverListener);
            legendEl.removeEventListener('mouseout', onMouseLeaveListener);
        };
    }, [legendEl]);
    const tooltipContent = legendData?.[legendVisibleIndex]?.tooltip;
    const tooltipContext = legendEl?.children[legendVisibleIndex];
    return ((0, jsx_runtime_1.jsxs)("div", { className: "calypso-uplot-chart-container", ref: uplotContainer, children: [isLoading && (0, jsx_runtime_1.jsx)(components_1.Spinner, { className: "calypso-uplot-chart-container__spinner" }), (0, jsx_runtime_1.jsx)(uplot_react_1.default, { data: data, onCreate: (chart) => (uplot.current = chart), options: options }), tooltipContent && tooltipContext && ((0, jsx_runtime_1.jsx)(popover_1.default, { isVisible: true, context: tooltipContext, position: "top", children: tooltipContent }))] }));
}
exports.default = UplotBarChart;
