// @automattic/calypso-analytics/dist/cjs/utils/current-user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setCurrentUser = exports.getCurrentUser = void 0;
const tslib_1 = require("tslib");
const hash_pii_1 = tslib_1.__importDefault(require("./hash-pii"));
/**
 * Module variables
 */
let _currentUser;
/**
 * Gets current user.
 * @returns Current user.
 */
function getCurrentUser() {
    return _currentUser;
}
exports.getCurrentUser = getCurrentUser;
/**
 * Sets current user, (stored in javascript memory).
 * @param currentUser the user data for the current user
 * @returns Current user.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function setCurrentUser(currentUser) {
    if (!currentUser.ID ||
        isNaN(parseInt(currentUser.ID, 10)) ||
        !currentUser.username ||
        !currentUser.email) {
        return; // Invalid user data.
    }
    _currentUser = {
        ID: parseInt(currentUser.ID, 10),
        username: currentUser.username,
        email: currentUser.email,
        localeSlug: currentUser.localeSlug,
        hashedPii: {
            ID: (0, hash_pii_1.default)(currentUser.ID),
            username: (0, hash_pii_1.default)(currentUser.username.toLowerCase().replace(/\s/g, '')),
            email: (0, hash_pii_1.default)(currentUser.email.toLowerCase().replace(/\s/g, '')),
        },
    };
    return _currentUser;
}
exports.setCurrentUser = setCurrentUser;
