// @automattic/components/dist/cjs/progress-bar/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const clsx_1 = tslib_1.__importDefault(require("clsx"));
const React = tslib_1.__importStar(require("react"));
const screen_reader_text_1 = tslib_1.__importDefault(require("../screen-reader-text"));
require("./style.scss");
class ProgressBar extends React.PureComponent {
    static defaultProps = {
        total: 100,
        compact: false,
        isPulsing: false,
        canGoBackwards: false,
    };
    static getDerivedStateFromProps(props, state) {
        return {
            allTimeMax: Math.max(state.allTimeMax, props.value),
        };
    }
    state = {
        allTimeMax: this.props.value,
    };
    getCompletionPercentage() {
        const percentage = Math.ceil(((this.props.canGoBackwards ? this.props.value : this.state.allTimeMax) /
            this.props.total) *
            100);
        // The percentage should not be allowed to be more than 100
        return Math.min(percentage, 100);
    }
    renderBar() {
        const { color, title, total, value, style } = this.props;
        let styles = { width: this.getCompletionPercentage() + '%' };
        if (color) {
            styles.backgroundColor = color;
        }
        if (style) {
            styles = { ...styles, ...style };
        }
        return ((0, jsx_runtime_1.jsx)("div", { "aria-valuemax": total, "aria-valuemin": 0, "aria-valuenow": value, "aria-label": "progress bar", className: "progress-bar__progress", role: "progressbar", style: styles, children: title && (0, jsx_runtime_1.jsx)(screen_reader_text_1.default, { children: title }) }));
    }
    render() {
        const classes = (0, clsx_1.default)(this.props.className, 'progress-bar', {
            'is-compact': this.props.compact,
            'is-pulsing': this.props.isPulsing,
        });
        return (0, jsx_runtime_1.jsx)("div", { className: classes, children: this.renderBar() });
    }
}
exports.default = ProgressBar;
