// @automattic/components/dist/cjs/gm-closure-notice/gm-closure-notice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GMClosureNotice = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const components_1 = require("@wordpress/components");
const date_1 = require("@wordpress/date");
const i18n_1 = require("@wordpress/i18n");
const react_i18n_1 = require("@wordpress/react-i18n");
require("./gm-closure-notice.scss");
const DATE_FORMAT_SHORT = 'F d';
const DATE_FORMAT_LONG = 'l, F jS h:i A';
function inBetween(start, end) {
    const date = new Date();
    return date >= start && date <= end;
}
function GMClosureNotice({ displayAt, closesAt, reopensAt, enabled }) {
    const { __ } = (0, react_i18n_1.useI18n)();
    if (!enabled) {
        return null;
    }
    const displayAtDate = new Date(displayAt);
    const closesAtDate = new Date(closesAt);
    const reopensAtDate = new Date(reopensAt);
    const MAIN_MESSAGES = {
        before: (0, i18n_1.sprintf)(
        /* translators: closes_at and reopens_at are dates */
        __('During this time, we will continue to provide support over email. If you need to get in touch with us submit a support request from this page, and we will get to it as fast as we can. Chat will re-open at %(reopens_at)s. Thank you for your understanding!', __i18n_text_domain__), {
            reopens_at: (0, date_1.format)(DATE_FORMAT_LONG, reopensAtDate),
        }),
        during: (0, i18n_1.sprintf)(
        /* translators:  reopens_at is a date */
        __('Chat will re-open at %(reopens_at)s. If you need to get in touch with us now, please submit a support request from this page. We will get to it as fast as we can. Thank you for your understanding!', __i18n_text_domain__), {
            reopens_at: (0, date_1.format)(DATE_FORMAT_LONG, reopensAtDate),
        }),
    };
    if (!inBetween(displayAtDate, reopensAtDate)) {
        return null;
    }
    const period = inBetween(closesAtDate, reopensAtDate) ? 'during' : 'before';
    const isSameMonth = reopensAtDate.getMonth() === closesAtDate.getMonth();
    const heading = (0, i18n_1.sprintf)(
    /* translators: closes and reopens are dates */
    __('Live chat will be closed from %(closes)s – %(reopens)s for the New Year’s holiday', __i18n_text_domain__), {
        closes: (0, date_1.format)(DATE_FORMAT_SHORT, closesAtDate),
        reopens: (0, date_1.format)(isSameMonth ? 'd' : DATE_FORMAT_SHORT, reopensAtDate),
    });
    return ((0, jsx_runtime_1.jsx)(components_1.Panel, { className: "a8c-components__gm-closure-notice", children: (0, jsx_runtime_1.jsx)(components_1.PanelBody, { initialOpen: period === 'during', title: heading, children: (0, jsx_runtime_1.jsx)(components_1.PanelRow, { children: MAIN_MESSAGES[period] }) }) }));
}
exports.GMClosureNotice = GMClosureNotice;
