// @automattic/components/dist/cjs/horizontal-bar-list/stories/index.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithRightItem = exports.WithLeftItem = exports.WithClick = exports.Indicated = exports.DefaultWithFooter = exports.Default = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const addon_actions_1 = require("@storybook/addon-actions");
const icons_1 = require("@wordpress/icons");
const _1 = tslib_1.__importDefault(require("../."));
const horizontal_bar_grid_item_1 = tslib_1.__importDefault(require("../horizontal-bar-grid-item"));
const stats_card_1 = tslib_1.__importDefault(require("../stats-card"));
require("./stories.scss");
exports.default = { title: 'packages/components/Horizontal bar list' };
const handleClick = (0, addon_actions_1.action)('click');
const data = [
    {
        label: 'Home page',
        value: 1000,
    },
    {
        label: 'The Lord of the Rings',
        value: 789,
    },
    {
        label: 'The Dark Knoght Trilogy',
        value: 512,
    },
    {
        label: 'The Ultimate Matrix Collection',
        value: 256,
    },
    {
        label: 'Ghost In The Shell',
        value: 110,
    },
    {
        label: 'Akira',
        value: 10,
    },
];
const HorizontalBarListVariations = (props) => {
    const { leftSideItem, renderRightSideItem, footerAction } = props;
    const testData = props.data || data;
    const barMaxValue = testData[0]?.value;
    return ((0, jsx_runtime_1.jsx)("div", { style: { width: '500px', margin: '0 auto' }, children: (0, jsx_runtime_1.jsx)(stats_card_1.default, { title: "Posts & Pages", footerAction: footerAction, children: (0, jsx_runtime_1.jsx)(_1.default, { data: testData, onClick: handleClick, children: testData.map((item, idx) => {
                    const variationProps = {
                        hasIndicator: props.hasIndicator && idx % 3, // omit every 3rd item from being indicated
                        onClick: props.onClick || null,
                        leftSideItem,
                        renderRightSideItem,
                    };
                    return ((0, jsx_runtime_1.jsx)(horizontal_bar_grid_item_1.default, { data: item, maxValue: barMaxValue, ...variationProps }, item?.id || idx));
                }) }) }) }));
};
const ImageSample = () => {
    return (0, jsx_runtime_1.jsx)("img", { src: "https://placekitten.com/20/20", alt: "sample" });
};
//TODO: convert to a component for mixing and matching actions.
const RedirectSample = () => {
    return ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(icons_1.Icon, { icon: icons_1.external, size: 18 }), (0, jsx_runtime_1.jsx)("span", { children: "View" })] }));
};
const Default = () => (0, jsx_runtime_1.jsx)(HorizontalBarListVariations, {});
exports.Default = Default;
const DefaultWithFooter = () => ((0, jsx_runtime_1.jsx)(HorizontalBarListVariations, { footerAction: { url: 'https://wordpress.com/', label: 'Show more' } }));
exports.DefaultWithFooter = DefaultWithFooter;
const Indicated = () => (0, jsx_runtime_1.jsx)(HorizontalBarListVariations, { hasIndicator: true });
exports.Indicated = Indicated;
const WithClick = () => (
// eslint-disable-next-line no-console
(0, jsx_runtime_1.jsx)(HorizontalBarListVariations, { onClick: () => console.log('I was clicked!') }));
exports.WithClick = WithClick;
const WithLeftItem = () => (0, jsx_runtime_1.jsx)(HorizontalBarListVariations, { leftSideItem: (0, jsx_runtime_1.jsx)(ImageSample, {}) });
exports.WithLeftItem = WithLeftItem;
const WithRightItem = () => ((0, jsx_runtime_1.jsx)(HorizontalBarListVariations, { renderRightSideItem: () => (0, jsx_runtime_1.jsx)(RedirectSample, {}) }));
exports.WithRightItem = WithRightItem;
