// i18n-calypso/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "I18N", {
  enumerable: true,
  get: function () {
    return _i18n.default;
  }
});
Object.defineProperty(exports, "I18NContext", {
  enumerable: true,
  get: function () {
    return _context.default;
  }
});
exports.isRtl = exports.getLocaleVariant = exports.getLocaleSlug = exports.getLocale = exports.emit = exports.default = exports.configure = exports.addTranslations = void 0;
Object.defineProperty(exports, "localize", {
  enumerable: true,
  get: function () {
    return _localize.default;
  }
});
exports.translate = exports.stateObserver = exports.state = exports.setLocale = exports.registerTranslateHook = exports.registerComponentUpdateHook = exports.reRenderTranslations = exports.on = exports.off = exports.numberFormat = void 0;
Object.defineProperty(exports, "useRtl", {
  enumerable: true,
  get: function () {
    return _rtl.useRtl;
  }
});
Object.defineProperty(exports, "useTranslate", {
  enumerable: true,
  get: function () {
    return _useTranslate.default;
  }
});
Object.defineProperty(exports, "withRtl", {
  enumerable: true,
  get: function () {
    return _rtl.withRtl;
  }
});
var _context = _interopRequireDefault(require("./context"));
var _defaultI18n = _interopRequireDefault(require("./default-i18n"));
var _i18n = _interopRequireDefault(require("./i18n"));
var _localize = _interopRequireDefault(require("./localize"));
var _rtl = require("./rtl");
var _useTranslate = _interopRequireDefault(require("./use-translate"));
var _default = exports.default = _defaultI18n.default; // Export the default instance's properties and bound methods for convenience
// These should be deprecated eventually, exposing only the default `i18n` instance
const numberFormat = exports.numberFormat = _defaultI18n.default.numberFormat.bind(_defaultI18n.default);
const translate = exports.translate = _defaultI18n.default.translate.bind(_defaultI18n.default);
const configure = exports.configure = _defaultI18n.default.configure.bind(_defaultI18n.default);
const setLocale = exports.setLocale = _defaultI18n.default.setLocale.bind(_defaultI18n.default);
const getLocale = exports.getLocale = _defaultI18n.default.getLocale.bind(_defaultI18n.default);
const getLocaleSlug = exports.getLocaleSlug = _defaultI18n.default.getLocaleSlug.bind(_defaultI18n.default);
const getLocaleVariant = exports.getLocaleVariant = _defaultI18n.default.getLocaleVariant.bind(_defaultI18n.default);
const isRtl = exports.isRtl = _defaultI18n.default.isRtl.bind(_defaultI18n.default);
const addTranslations = exports.addTranslations = _defaultI18n.default.addTranslations.bind(_defaultI18n.default);
const reRenderTranslations = exports.reRenderTranslations = _defaultI18n.default.reRenderTranslations.bind(_defaultI18n.default);
const registerComponentUpdateHook = exports.registerComponentUpdateHook = _defaultI18n.default.registerComponentUpdateHook.bind(_defaultI18n.default);
const registerTranslateHook = exports.registerTranslateHook = _defaultI18n.default.registerTranslateHook.bind(_defaultI18n.default);
const state = exports.state = _defaultI18n.default.state;
const stateObserver = exports.stateObserver = _defaultI18n.default.stateObserver;
const on = exports.on = _defaultI18n.default.on.bind(_defaultI18n.default);
const off = exports.off = _defaultI18n.default.off.bind(_defaultI18n.default);
const emit = exports.emit = _defaultI18n.default.emit.bind(_defaultI18n.default);
