// i18n-calypso/dist/cjs/rtl.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useRtl = useRtl;
exports.withRtl = void 0;
var _compose = require("@wordpress/compose");
var _react = require("react");
var _useSubscription = require("use-subscription");
var _context = _interopRequireDefault(require("./context"));
var _jsxRuntime = require("react/jsx-runtime");
function useRtl() {
  const i18n = (0, _react.useContext)(_context.default);
  // Subscription object (adapter) for the `useSubscription` hook
  const RtlSubscription = (0, _react.useMemo)(() => ({
    getCurrentValue() {
      return i18n.isRtl();
    },
    subscribe(callback) {
      i18n.on('change', callback);
      return () => i18n.off('change', callback);
    }
  }), [i18n]);
  return (0, _useSubscription.useSubscription)(RtlSubscription);
}
const withRtl = exports.withRtl = (0, _compose.createHigherOrderComponent)(WrappedComponent => /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const isRtl = useRtl();
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(WrappedComponent, {
    ...props,
    isRtl: isRtl,
    ref: ref
  });
}), 'WithRTL');
