// @automattic/components/dist/cjs/standalone-comparison-grid/index.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Normal = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const components_1 = require("@wordpress/components");
const _1 = require(".");
require("./stories.scss");
exports.default = { title: 'packages/components/StandaloneComparisonGrid' };
const columns = [
    {
        title: 'Import',
        features: [
            'Import your pages and posts',
            'Import your uploaded images and files',
            'Import your users**',
            'Import your menus** and FSE templates**',
        ],
        introCopy: 'Import your site content, without themes, customizations, or plugins. You will probably have to do some work to get things looking exactly the same.',
        controls: ((0, jsx_runtime_1.jsx)(components_1.Button, { variant: "primary", onClick: () => { }, children: "Import my website content" })),
    },
    {
        title: 'Migrate',
        features: [
            'Import your pages and posts',
            'Import your uploaded images and files',
            'Import your users',
            'Import your menus and FSE templates',
            'Import your theme and exact layout',
            'Import your plugins and existing setup',
            'Custom blocks from all plugins keep working',
            'Import custom files and PHP code',
            'Import content from page builders like Elementor, Divi, or WP Bakery',
            'Import all WooCommerce products and orders',
        ],
        introCopy: 'Available with the Creator Plan from $NN/month',
        controls: ((0, jsx_runtime_1.jsx)(components_1.Button, { variant: "primary", onClick: () => { }, children: "Migrate my website" })),
        action: () => { },
    },
];
const GridVariations = () => ((0, jsx_runtime_1.jsxs)(_1.StandAloneComparisonGrid, { children: [(0, jsx_runtime_1.jsx)(_1.Column, { ...columns[0] }), (0, jsx_runtime_1.jsx)(_1.Column, { ...columns[1] })] }));
const Normal = () => (0, jsx_runtime_1.jsx)(GridVariations, {});
exports.Normal = Normal;
