// @automattic/components/dist/esm/jetpack-upsell-card/stories/index.stories.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import JetpackUpsellCard from '..';
export default { title: 'packages/components/Jetpack Upsell Card' };
const upgradeUrls = {
    backup: 'https://jetpack.com',
    boost: 'https://jetpack.com',
    search: 'https://jetpack.com',
    security: 'https://jetpack.com',
    social: 'https://jetpack.com',
    video: 'https://jetpack.com',
};
const UpsellSection = (props) => (_jsx("div", { className: "jetpack-upsell-section-story", style: { margin: 'auto', maxWidth: '1000px' }, children: _jsx(JetpackUpsellCard, { purchasedProducts: [], upgradeUrls: upgradeUrls, ...props }) }));
export const Default = () => _jsx(UpsellSection, {});
export const WithACustomSiteSlug = () => _jsx(UpsellSection, { siteSlug: "YourCustomDomain.com" });
export const WithNoUpgradeUrls = () => (_jsxs(_Fragment, { children: [_jsx(UpsellSection, { upgradeUrls: {} }), _jsx("p", { children: "Nothing should be rendered for this story." })] }));
