// @automattic/components/dist/cjs/app-promo-card/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const i18n_utils_1 = require("@automattic/i18n-utils");
const clsx_1 = tslib_1.__importDefault(require("clsx"));
const i18n_calypso_1 = require("i18n-calypso");
const qrcode_react_1 = tslib_1.__importDefault(require("qrcode.react"));
require("./style.scss");
const icon_woo_png_1 = tslib_1.__importDefault(require("./images/icon-woo.png"));
const qr_code_woo_png_1 = tslib_1.__importDefault(require("./images/qr-code-woo.png"));
const store_apple_png_1 = tslib_1.__importDefault(require("./images/store-apple.png"));
const store_google_png_1 = tslib_1.__importDefault(require("./images/store-google.png"));
const svg_icons_1 = require("./svg-icons");
// Slugs as used by Jetpack Redirects.
// See https://jetpack.com/redirect for current URLs.
// Two _QRCode constants are included for reference only.
const REDIRECT_SLUGS = {
    jetpackApple: 'calypso-stats-mobile-cta-jetpack-apple-badge',
    jetpackGoogle: 'calypso-stats-mobile-cta-jetpack-google-badge',
    wooA8C: 'calypso-stats-mobile-cta-woo-link',
    wooApple: 'calypso-stats-mobile-cta-woo-apple-badge',
    wooGoogle: 'calypso-stats-mobile-cta-woo-google-badge',
    wooQRCode: 'calypso-stats-mobile-cta-woo-qrcode',
};
// Generate a Jetpack Redirect URL based on the provided slug.
// Similar to getRedirectUrl from '@automattic/jetpack-components'.
function getRedirectUrl(key) {
    // Confirm requested slug is valid.
    if (!REDIRECT_SLUGS.hasOwnProperty(key)) {
        return null;
    }
    // Return redirect URL.
    return 'https://jetpack.com/redirect/?source=' + REDIRECT_SLUGS[key];
}
// For relaying click events to the caller.
const CLICK_EVENTS = {
    jetpackClickA8C: 'jetpack-click-a8c',
    jetpackClickApple: 'jetpack-click-apple',
    jetpackClickGoogle: 'jetpack-click-google',
    wooClickA8C: 'woo-click-a8c',
    wooClickApple: 'woo-click-apple',
    wooClickGoogle: 'woo-click-google',
};
function AppPromoCard({ className, isWoo, clickHandler }) {
    const translate = (0, i18n_calypso_1.useTranslate)();
    const localizeUrl = (0, i18n_utils_1.useLocalizeUrl)();
    // Basic user agent testing so we can show app store badges on moble.
    const userAgent = window.navigator.userAgent.toLowerCase();
    const isApple = userAgent.includes('iphone') || userAgent.includes('ipad');
    const isGoogle = userAgent.includes('android');
    const onClickHandler = (eventName) => {
        clickHandler?.(eventName);
    };
    // Determines message text based on mobile, tablet, or Desktop.
    const getMessage = () => {
        // The mobile use case. If Apple or Google mobile device, Jetpack or Woo.
        if (isApple || isGoogle) {
            return isWoo
                ? translate('Check your stats on-the-go and get real-time notifications with the Woo mobile app.')
                : translate('Check your stats on-the-go and get real-time notifications with the Jetpack mobile app.');
        }
        // Using useTranslate() with interpolation to set up the linked message.
        // https://wpcalypso.wordpress.com/devdocs/packages/i18n-calypso/README.md
        if (isWoo) {
            return translate('Visit {{a}}woo.com/mobile{{/a}} or scan the QR code to download the WooCommerce mobile app.', {
                components: {
                    a: ((0, jsx_runtime_1.jsx)("a", { className: "app-promo-card__link app-promo-card__woo-link", href: getRedirectUrl('wooA8C') ?? 'https://woo.com/mobile', onClick: () => onClickHandler(CLICK_EVENTS.wooClickA8C) })),
                },
            });
        }
        return translate('Visit {{a}}wp.com/app{{/a}} or scan the QR code to download the Jetpack mobile app.', {
            components: {
                a: ((0, jsx_runtime_1.jsx)("a", { className: "app-promo-card__link app-promo-card__jetpack-link", href: localizeUrl('https://apps.wordpress.com/get/?campaign=calypso-stats-promo'), onClick: () => onClickHandler(CLICK_EVENTS.jetpackClickA8C) })),
            },
        });
    };
    // Returns store badges on mobile (including tablets) and QR codes on the Desktop.
    const getPromoImage = () => {
        const fallbackLink = isWoo ? 'https://woo.com/mobile' : 'https://wp.com/app';
        if (isApple) {
            const appStoreLink = isWoo ? getRedirectUrl('wooApple') : getRedirectUrl('jetpackApple');
            const tracksEventName = isWoo ? CLICK_EVENTS.wooClickApple : CLICK_EVENTS.jetpackClickApple;
            return ((0, jsx_runtime_1.jsx)("a", { href: appStoreLink ?? fallbackLink, onClick: () => onClickHandler(tracksEventName), children: (0, jsx_runtime_1.jsx)("img", { className: "app-promo-card__store-badge", src: store_apple_png_1.default, alt: "Badge for the Apple App Store" }) }));
        }
        if (isGoogle) {
            const appStoreLink = isWoo
                ? getRedirectUrl('wooGoogle')
                : getRedirectUrl('jetpackGoogle');
            const tracksEventName = isWoo ? CLICK_EVENTS.wooClickGoogle : CLICK_EVENTS.jetpackClickGoogle;
            return ((0, jsx_runtime_1.jsx)("a", { href: appStoreLink ?? fallbackLink, onClick: () => onClickHandler(tracksEventName), children: (0, jsx_runtime_1.jsx)("img", { className: "app-promo-card__store-badge", src: store_google_png_1.default, alt: "Badge for the Google Play Store" }) }));
        }
        return isWoo ? ((0, jsx_runtime_1.jsx)("img", { className: "app-promo-card__qr-code", src: qr_code_woo_png_1.default, alt: "QR Code for Woo mobile app" })) : ((0, jsx_runtime_1.jsx)(qrcode_react_1.default, { className: "app-promo-card__qr-code", value: localizeUrl('https://apps.wordpress.com/get/?campaign=calypso-qrcode-apps'), size: 300 }));
    };
    return ((0, jsx_runtime_1.jsxs)("div", { className: (0, clsx_1.default)('app-promo-card', className ?? null), children: [(0, jsx_runtime_1.jsxs)("div", { className: "app-promo-card__lhs", children: [(0, jsx_runtime_1.jsxs)("div", { className: "app-promo-card__icons", children: [isWoo && ((0, jsx_runtime_1.jsx)("img", { className: "app-promo-card__woo-icon", src: icon_woo_png_1.default, alt: "Icon for the Woo mobile app" })), !isWoo && (0, jsx_runtime_1.jsx)(svg_icons_1.WordPressJetpackSVG, {})] }), (0, jsx_runtime_1.jsx)("p", { className: "app-promo-card__title", children: isWoo
                            ? translate('Bring your Store stats with you using the Woo mobile app')
                            : translate('Bring your stats with you using the Jetpack mobile app') }), (0, jsx_runtime_1.jsx)("p", { className: "app-promo-card__message", children: getMessage() })] }), (0, jsx_runtime_1.jsx)("div", { className: "app-promo-card__rhs", children: getPromoImage() })] }));
}
exports.default = AppPromoCard;
