// @automattic/components/dist/cjs/jetpack-upsell-card/stories/index.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithNoUpgradeUrls = exports.WithACustomSiteSlug = exports.Default = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const __1 = tslib_1.__importDefault(require(".."));
exports.default = { title: 'packages/components/Jetpack Upsell Card' };
const upgradeUrls = {
    backup: 'https://jetpack.com',
    boost: 'https://jetpack.com',
    search: 'https://jetpack.com',
    security: 'https://jetpack.com',
    social: 'https://jetpack.com',
    video: 'https://jetpack.com',
};
const UpsellSection = (props) => ((0, jsx_runtime_1.jsx)("div", { className: "jetpack-upsell-section-story", style: { margin: 'auto', maxWidth: '1000px' }, children: (0, jsx_runtime_1.jsx)(__1.default, { purchasedProducts: [], upgradeUrls: upgradeUrls, ...props }) }));
const Default = () => (0, jsx_runtime_1.jsx)(UpsellSection, {});
exports.Default = Default;
const WithACustomSiteSlug = () => (0, jsx_runtime_1.jsx)(UpsellSection, { siteSlug: "YourCustomDomain.com" });
exports.WithACustomSiteSlug = WithACustomSiteSlug;
const WithNoUpgradeUrls = () => ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(UpsellSection, { upgradeUrls: {} }), (0, jsx_runtime_1.jsx)("p", { children: "Nothing should be rendered for this story." })] }));
exports.WithNoUpgradeUrls = WithNoUpgradeUrls;
