// @woocommerce/components/build-module/tour-kit/index.js
/**
 * External dependencies
 */
import { createElement } from '@wordpress/element';
import TourKit from '@automattic/tour-kit';
/**
 * Internal dependencies
 */
import TourKitStep from './components/step';
const defaultOptions = {
    effects: {
        spotlight: {
            interactivity: { enabled: true, rootElementSelector: '#wpwrap' },
        },
        arrowIndicator: true,
        liveResize: {
            mutation: true,
            resize: true,
            rootElementSelector: '#wpwrap',
        },
    },
};
const WooTourKit = ({ config }) => {
    return (createElement(TourKit, { __temp__className: 'woocommerce-tour-kit', config: {
            options: {
                ...defaultOptions,
                ...config.options,
            },
            ...config,
            renderers: {
                tourStep: TourKitStep,
                // Disable minimize feature for woo tour kit.
                tourMinimized: () => null,
            },
        } }));
};
export default WooTourKit;
