// @woocommerce/components/build/search/autocompleters/countries.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * External dependencies
 */
const html_entities_1 = require("@wordpress/html-entities");
const element_1 = require("@wordpress/element");
const api_fetch_1 = __importDefault(require("@wordpress/api-fetch"));
/**
 * Internal dependencies
 */
const utils_1 = require("./utils");
const flag_1 = __importDefault(require("../../flag"));
/**
 * A raw completer option.
 *
 * @typedef {*} CompleterOption
 */
/**
 * @callback FnGetOptions
 *
 * @return {(CompleterOption[]|Promise.<CompleterOption[]>)} The completer options or a promise for them.
 */
/**
 * @callback FnGetOptionKeywords
 * @param {CompleterOption} option a completer option.
 *
 * @return {string[]} list of key words to search.
 */
/**
 * @callback FnIsOptionDisabled
 * @param {CompleterOption} option a completer option.
 *
 * @return {string[]} whether or not the given option is disabled.
 */
/**
 * @callback FnGetOptionLabel
 * @param {CompleterOption} option a completer option.
 *
 * @return {(string|Array.<(string|Node)>)} list of react components to render.
 */
/**
 * @callback FnAllowContext
 * @param {string} before the string before the auto complete trigger and query.
 * @param {string} after  the string after the autocomplete trigger and query.
 *
 * @return {boolean} true if the completer can handle.
 */
/**
 * @typedef {Object} OptionCompletion
 * @property {'insert-at-caret'|'replace'} action the intended placement of the completion.
 * @property {OptionCompletionValue}       value  the completion value.
 */
/**
 * A completion value.
 *
 * @typedef {(string|WPElement|Object)} OptionCompletionValue
 */
/**
 * @callback FnGetOptionCompletion
 * @param {CompleterOption} value the value of the completer option.
 * @param {string}          query the text value of the autocomplete query.
 *
 * @return {(OptionCompletion|OptionCompletionValue)} the completion for the given option. If an
 * 													   OptionCompletionValue is returned, the
 * 													   completion action defaults to `insert-at-caret`.
 */
/**
 * @typedef {Object} WPCompleter
 * @property {string}                           name                a way to identify a completer, useful for selective overriding.
 * @property {?string}                          className           A class to apply to the popup menu.
 * @property {string}                           triggerPrefix       the prefix that will display the menu.
 * @property {(CompleterOption[]|FnGetOptions)} options             the completer options or a function to get them.
 * @property {?FnGetOptionKeywords}             getOptionKeywords   get the keywords for a given option.
 * @property {?FnIsOptionDisabled}              isOptionDisabled    get whether or not the given option is disabled.
 * @property {FnGetOptionLabel}                 getOptionLabel      get the label for a given option.
 * @property {?FnAllowContext}                  allowContext        filter the context under which the autocomplete activates.
 * @property {FnGetOptionCompletion}            getOptionCompletion get the completion associated with a given option.
 */
// Cache countries to avoid repeated requests.
let allCountries = null;
/**
 * A country completer.
 * See https://github.com/WordPress/gutenberg/tree/master/packages/components/src/autocomplete#the-completer-interface
 *
 * @type {WPCompleter}
 */
exports.default = {
    name: 'countries',
    className: 'woocommerce-search__country-result',
    isDebounced: true,
    options() {
        // Returned cached countries if we've already received them.
        if (allCountries) {
            return Promise.resolve(allCountries);
        }
        // Make the request for country data.
        return (0, api_fetch_1.default)({ path: '/wc-analytics/data/countries' }).then((result) => {
            // Cache the response.
            allCountries = result;
            return allCountries;
        });
    },
    getOptionIdentifier(country) {
        return country.code;
    },
    getSearchExpression(query) {
        return '^' + query;
    },
    getOptionKeywords(country) {
        return [country.code, (0, html_entities_1.decodeEntities)(country.name)];
    },
    getOptionLabel(country, query) {
        const name = (0, html_entities_1.decodeEntities)(country.name);
        const match = (0, utils_1.computeSuggestionMatch)(name, query) || {};
        return ((0, element_1.createElement)(element_1.Fragment, null,
            (0, element_1.createElement)(flag_1.default, { key: "thumbnail", className: "woocommerce-search__result-thumbnail", code: country.code, size: 18, hideFromScreenReader: true }),
            (0, element_1.createElement)("span", { key: "name", className: "woocommerce-search__result-name", "aria-label": name }, query ? ((0, element_1.createElement)(element_1.Fragment, null,
                match.suggestionBeforeMatch,
                (0, element_1.createElement)("strong", { className: "components-form-token-field__suggestion-match" }, match.suggestionMatch),
                match.suggestionAfterMatch)) : (name))));
    },
    // This is slightly different than gutenberg/Autocomplete, we don't support different methods
    // of replace/insertion, so we can just return the value.
    getOptionCompletion(country) {
        const value = {
            key: country.code,
            label: (0, html_entities_1.decodeEntities)(country.name),
        };
        return value;
    },
};
