// @woocommerce/components/build/search/autocompleters/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.variations = exports.variableProduct = exports.usernames = exports.taxes = exports.product = exports.orders = exports.emails = exports.downloadIps = exports.customers = exports.coupons = exports.countries = exports.productCategory = exports.attributes = void 0;
/**
 * Export all autocompleters
 */
var attributes_1 = require("./attributes");
Object.defineProperty(exports, "attributes", { enumerable: true, get: function () { return __importDefault(attributes_1).default; } });
var categories_1 = require("./categories");
Object.defineProperty(exports, "productCategory", { enumerable: true, get: function () { return __importDefault(categories_1).default; } });
var countries_1 = require("./countries");
Object.defineProperty(exports, "countries", { enumerable: true, get: function () { return __importDefault(countries_1).default; } });
var coupons_1 = require("./coupons");
Object.defineProperty(exports, "coupons", { enumerable: true, get: function () { return __importDefault(coupons_1).default; } });
var customers_1 = require("./customers");
Object.defineProperty(exports, "customers", { enumerable: true, get: function () { return __importDefault(customers_1).default; } });
var download_ips_1 = require("./download-ips");
Object.defineProperty(exports, "downloadIps", { enumerable: true, get: function () { return __importDefault(download_ips_1).default; } });
var emails_1 = require("./emails");
Object.defineProperty(exports, "emails", { enumerable: true, get: function () { return __importDefault(emails_1).default; } });
var orders_1 = require("./orders");
Object.defineProperty(exports, "orders", { enumerable: true, get: function () { return __importDefault(orders_1).default; } });
var product_1 = require("./product");
Object.defineProperty(exports, "product", { enumerable: true, get: function () { return __importDefault(product_1).default; } });
var taxes_1 = require("./taxes");
Object.defineProperty(exports, "taxes", { enumerable: true, get: function () { return __importDefault(taxes_1).default; } });
var usernames_1 = require("./usernames");
Object.defineProperty(exports, "usernames", { enumerable: true, get: function () { return __importDefault(usernames_1).default; } });
var variable_product_1 = require("./variable-product");
Object.defineProperty(exports, "variableProduct", { enumerable: true, get: function () { return __importDefault(variable_product_1).default; } });
var variations_1 = require("./variations");
Object.defineProperty(exports, "variations", { enumerable: true, get: function () { return __importDefault(variations_1).default; } });
