// @automattic/components/dist/cjs/upsell-menu-group/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpsellMenuGroup = exports.UpsellIcon = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const styled_1 = tslib_1.__importDefault(require("@emotion/styled"));
const gridicon_1 = tslib_1.__importDefault(require("../gridicon"));
const Row = styled_1.default.div({
    display: 'flex',
    flexDirection: 'row',
    paddingTop: 8,
    paddingBottom: 8,
    gap: 8,
});
const Content = styled_1.default.div({
    display: 'flex',
    flexDirection: 'column',
    gap: 8,
    width: 210,
    whiteSpace: 'break-spaces',
    maxWidth: '80vw',
});
const BackgroundIcon = styled_1.default.div({
    backgroundColor: 'var(--color-accent)',
    width: 16,
    height: 16,
    borderRadius: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: 1,
});
const StarIcon = (0, styled_1.default)(gridicon_1.default)({
    color: '#FFF',
});
function UpsellIcon() {
    return ((0, jsx_runtime_1.jsx)(BackgroundIcon, { children: (0, jsx_runtime_1.jsx)(StarIcon, { icon: "star", size: 10 }) }));
}
exports.UpsellIcon = UpsellIcon;
function UpsellMenuGroup(props) {
    const { icon = (0, jsx_runtime_1.jsx)(UpsellIcon, {}), children, ...rest } = props;
    return ((0, jsx_runtime_1.jsxs)(Row, { ...rest, children: [icon, (0, jsx_runtime_1.jsx)(Content, { children: children })] }));
}
exports.UpsellMenuGroup = UpsellMenuGroup;
