// @automattic/components/dist/esm/post-stats-card/stories/index.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import PostStatsCard from '../';
export default { title: 'packages/components/Post Stats Card' };
const PostStatsCardVariations = (props) => (_jsx(PostStatsCard, { commentCount: 7, heading: "Latest post", likeCount: 83, viewCount: 912, post: {
        date: '2022-12-07T14:24:48+00:00',
        post_thumbnail: 'https://images.unsplash.com/photo-1670510521067-2e94183f9803',
        title: 'The moody trail',
    }, ...props }));
export const Default = () => (_jsx("div", { className: "post-stats-card-story", style: { margin: 'auto', maxWidth: '700px' }, children: _jsx(PostStatsCardVariations, {}) }));
export const FullWidth = () => (_jsx("div", { className: "post-stats-card-story", children: _jsx(PostStatsCardVariations, {}) }));
export const WithoutAThumbnail = () => (_jsx("div", { className: "post-stats-card-story", children: _jsx(PostStatsCardVariations, { post: {
            date: '2022-12-07T14:24:48+00:00',
            post_thumbnail: null,
            title: 'The moody trail',
        } }) }));
