// @woocommerce/components/build/timeline/timeline-item.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * External dependencies
 */
const classnames_1 = __importDefault(require("classnames"));
const date_1 = require("@wordpress/date");
const prop_types_1 = __importDefault(require("prop-types"));
const element_1 = require("@wordpress/element");
const TimelineItem = (props) => {
    const { item, className, clockFormat } = props;
    const itemClassName = (0, classnames_1.default)('woocommerce-timeline-item', className);
    const itemTimeString = (0, date_1.format)(clockFormat, item.date);
    return ((0, element_1.createElement)("li", { className: itemClassName },
        (0, element_1.createElement)("div", { className: 'woocommerce-timeline-item__top-border' }),
        (0, element_1.createElement)("div", { className: 'woocommerce-timeline-item__title' },
            (0, element_1.createElement)("div", { className: 'woocommerce-timeline-item__headline' },
                item.icon,
                (0, element_1.createElement)("span", null, item.headline)),
            (0, element_1.createElement)("span", { className: 'woocommerce-timeline-item__timestamp' }, item.hideTimestamp || false ? null : itemTimeString)),
        (0, element_1.createElement)("div", { className: 'woocommerce-timeline-item__body' }, (item.body || []).map((bodyItem, index) => ((0, element_1.createElement)("span", { key: `timeline-item-body-${index}` }, bodyItem))))));
};
TimelineItem.propTypes = {
    /**
     * Additional CSS classes.
     */
    className: prop_types_1.default.string,
    /**
     * An array of list items.
     */
    item: prop_types_1.default.shape({
        /**
         * Date for the timeline item.
         */
        date: prop_types_1.default.instanceOf(Date).isRequired,
        /**
         * Icon for the Timeline item.
         */
        icon: prop_types_1.default.element.isRequired,
        /**
         * Headline displayed for the list item.
         */
        headline: prop_types_1.default.oneOfType([prop_types_1.default.element, prop_types_1.default.string])
            .isRequired,
        /**
         * Body displayed for the list item.
         */
        body: prop_types_1.default.arrayOf(prop_types_1.default.oneOfType([prop_types_1.default.element, prop_types_1.default.string])),
        /**
         * Allows users to toggle the timestamp on or off.
         */
        hideTimestamp: prop_types_1.default.bool,
        /**
         * The PHP clock format string used to format times, see php.net/date.
         */
        clockFormat: prop_types_1.default.string,
    }).isRequired,
};
TimelineItem.defaultProps = {
    className: '',
    item: {},
};
exports.default = TimelineItem;
