// @woocommerce/components/build/tour-kit/index.js
"use strict";
/**
 * External dependencies
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const element_1 = require("@wordpress/element");
const tour_kit_1 = __importDefault(require("@automattic/tour-kit"));
/**
 * Internal dependencies
 */
const step_1 = __importDefault(require("./components/step"));
const defaultOptions = {
    effects: {
        spotlight: {
            interactivity: { enabled: true, rootElementSelector: '#wpwrap' },
        },
        arrowIndicator: true,
        liveResize: {
            mutation: true,
            resize: true,
            rootElementSelector: '#wpwrap',
        },
    },
};
const WooTourKit = ({ config }) => {
    return ((0, element_1.createElement)(tour_kit_1.default, { __temp__className: 'woocommerce-tour-kit', config: Object.assign(Object.assign({ options: Object.assign(Object.assign({}, defaultOptions), config.options) }, config), { renderers: {
                tourStep: step_1.default,
                // Disable minimize feature for woo tour kit.
                tourMinimized: () => null,
            } }) }));
};
exports.default = WooTourKit;
