// @automattic/components/dist/cjs/spinner/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Spinner = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const clsx_1 = tslib_1.__importDefault(require("clsx"));
const react_1 = require("react");
require("./style.scss");
class Spinner extends react_1.PureComponent {
    static defaultProps = {
        size: 20,
        baseClassName: 'spinner',
    };
    render() {
        const className = (0, clsx_1.default)(this.props.baseClassName, this.props.className);
        const style = {
            width: this.props.size,
            height: this.props.size,
            fontSize: this.props.size, // allows border-width to be specified in em units
        };
        return ((0, jsx_runtime_1.jsx)("div", { className: className, children: (0, jsx_runtime_1.jsx)("div", { className: "spinner__outer", style: style, children: (0, jsx_runtime_1.jsx)("div", { className: "spinner__inner" }) }) }));
    }
}
exports.Spinner = Spinner;
