// @automattic/tour-kit/dist/cjs/variants/wpcom/storybook/index.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpotlightWithRating = exports.DefaultWithRating = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const element_1 = require("@wordpress/element");
const __1 = tslib_1.__importDefault(require(".."));
exports.default = { title: 'wpcom-tour-kit' };
const References = () => {
    return ((0, jsx_runtime_1.jsx)("div", { className: "storybook__tourkit-references", children: (0, jsx_runtime_1.jsxs)("div", { className: "storybook__tourkit-references-container", children: [(0, jsx_runtime_1.jsx)("div", { className: "storybook__tourkit-references-a", children: (0, jsx_runtime_1.jsx)("p", { children: "Reference A" }) }), (0, jsx_runtime_1.jsx)("div", { className: "storybook__tourkit-references-b", children: (0, jsx_runtime_1.jsx)("p", { children: "Reference B" }) }), (0, jsx_runtime_1.jsx)("div", { className: "storybook__tourkit-references-c", children: (0, jsx_runtime_1.jsx)("p", { children: "Reference C" }) }), (0, jsx_runtime_1.jsx)("div", { className: "storybook__tourkit-references-d", children: (0, jsx_runtime_1.jsx)("p", { children: "Reference D" }) })] }) }));
};
const Tour = ({ onClose, options }) => {
    const config = {
        steps: [
            {
                referenceElements: {
                    desktop: '.storybook__tourkit-references-a',
                    mobile: '.storybook__tourkit-references-a',
                },
                meta: {
                    heading: 'Laura 1',
                    descriptions: {
                        desktop: 'Lorem ipsum dolor sit amet.',
                        mobile: 'Lorem ipsum dolor sit amet.',
                    },
                },
            },
            {
                referenceElements: {
                    desktop: '.storybook__tourkit-references-b',
                    mobile: '.storybook__tourkit-references-b',
                },
                meta: {
                    heading: 'Laura 2',
                    descriptions: {
                        desktop: 'Lorem ipsum dolor sit amet.',
                        mobile: 'Lorem ipsum dolor sit amet.',
                    },
                },
            },
            {
                referenceElements: {
                    desktop: '.storybook__tourkit-references-c',
                    mobile: '.storybook__tourkit-references-c',
                },
                meta: {
                    heading: 'Laura 3',
                    descriptions: {
                        desktop: 'Lorem ipsum dolor sit amet.',
                        mobile: 'Lorem ipsum dolor sit amet.',
                    },
                },
            },
            {
                referenceElements: {
                    desktop: '.storybook__tourkit-references-d',
                    mobile: '.storybook__tourkit-references-d',
                },
                meta: {
                    heading: 'Laura 4',
                    descriptions: {
                        desktop: 'Lorem ipsum dolor sit amet.',
                        mobile: 'Lorem ipsum dolor sit amet.',
                    },
                },
            },
        ],
        closeHandler: onClose,
        options: {
            classNames: ['mytour'],
            ...options,
        },
    };
    return (0, jsx_runtime_1.jsx)(__1.default, { config: config });
};
const StoryTour = ({ options = {} }) => {
    const [showTour, setShowTour] = (0, element_1.useState)(false);
    return ((0, jsx_runtime_1.jsxs)("div", { className: "storybook__tourkit", children: [(0, jsx_runtime_1.jsx)(References, {}), !showTour && (0, jsx_runtime_1.jsx)("button", { onClick: () => setShowTour(true), children: "Start Tour" }), showTour && (0, jsx_runtime_1.jsx)(Tour, { onClose: () => setShowTour(false), options: options })] }));
};
const DefaultWithRating = () => ((0, jsx_runtime_1.jsx)(StoryTour, { options: { effects: { arrowIndicator: false }, tourRating: { enabled: true } } }));
exports.DefaultWithRating = DefaultWithRating;
const SpotlightWithRating = () => ((0, jsx_runtime_1.jsx)(StoryTour, { options: { effects: { spotlight: {}, arrowIndicator: false }, tourRating: { enabled: true } } }));
exports.SpotlightWithRating = SpotlightWithRating;
