// @woocommerce/components/build-module/summary/placeholder.js
/**
 * External dependencies
 */
import { createElement, Component } from '@wordpress/element';
import classnames from 'classnames';
import { range } from 'lodash';
import PropTypes from 'prop-types';
import { withViewportMatch } from '@wordpress/viewport';
/**
 * Internal dependencies
 */
import { getHasItemsClass } from './utils';
export const SummaryNumberPlaceholder = ({ className }) => (createElement("li", { "data-testid": "summary-placeholder", className: classnames('woocommerce-summary__item-container is-placeholder', className) },
    createElement("div", { className: "woocommerce-summary__item" },
        createElement("div", { className: "woocommerce-summary__item-label" }),
        createElement("div", { className: "woocommerce-summary__item-data" },
            createElement("div", { className: "woocommerce-summary__item-value" }),
            createElement("div", { className: "woocommerce-summary__item-delta" })))));
/**
 * `SummaryListPlaceholder` behaves like `SummaryList` but displays placeholder summary items instead of data.
 * This can be used while loading data.
 */
class SummaryListPlaceholder extends Component {
    render() {
        const { isDropdownBreakpoint } = this.props;
        const numberOfItems = isDropdownBreakpoint
            ? 1
            : this.props.numberOfItems;
        const hasItemsClass = getHasItemsClass(numberOfItems);
        const classes = classnames('woocommerce-summary', {
            [hasItemsClass]: !isDropdownBreakpoint,
            'is-placeholder': true,
        });
        return (createElement("ul", { className: classes, "aria-hidden": "true" }, range(numberOfItems).map((i) => {
            return createElement(SummaryNumberPlaceholder, { key: i });
        })));
    }
}
SummaryListPlaceholder.propTypes = {
    /**
     * An integer with the number of summary items to display.
     */
    numberOfItems: PropTypes.number.isRequired,
};
SummaryListPlaceholder.defaultProps = {
    numberOfRows: 5,
};
export default withViewportMatch({
    isDropdownBreakpoint: '< large',
})(SummaryListPlaceholder);
