// @wordpress/components/build-module/range-control/input-range.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { forwardRef } from '@wordpress/element';
/**
 * Internal dependencies
 */

import { InputRange as BaseInputRange } from './styles/range-control-styles';
import { useDebouncedHoverInteraction } from './utils';

const noop = () => {};

function InputRange(props, ref) {
  const {
    describedBy,
    label,
    onHideTooltip = noop,
    onMouseLeave = noop,
    onMouseMove = noop,
    onShowTooltip = noop,
    value,
    ...otherProps
  } = props;
  const hoverInteractions = useDebouncedHoverInteraction({
    onHide: onHideTooltip,
    onMouseLeave,
    onMouseMove,
    onShow: onShowTooltip
  });
  return createElement(BaseInputRange, _extends({}, otherProps, hoverInteractions, {
    "aria-describedby": describedBy,
    "aria-label": label,
    "aria-hidden": false,
    ref: ref,
    tabIndex: 0,
    type: "range",
    value: value
  }));
}

const ForwardedComponent = forwardRef(InputRange);
export default ForwardedComponent;
