// @automattic/tour-kit/dist/esm/variants/wpcom/storybook/index.stories.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from '@wordpress/element';
import WpcomTourKit from '..';
export default { title: 'wpcom-tour-kit' };
const References = () => {
    return (_jsx("div", { className: "storybook__tourkit-references", children: _jsxs("div", { className: "storybook__tourkit-references-container", children: [_jsx("div", { className: "storybook__tourkit-references-a", children: _jsx("p", { children: "Reference A" }) }), _jsx("div", { className: "storybook__tourkit-references-b", children: _jsx("p", { children: "Reference B" }) }), _jsx("div", { className: "storybook__tourkit-references-c", children: _jsx("p", { children: "Reference C" }) }), _jsx("div", { className: "storybook__tourkit-references-d", children: _jsx("p", { children: "Reference D" }) })] }) }));
};
const Tour = ({ onClose, options }) => {
    const config = {
        steps: [
            {
                referenceElements: {
                    desktop: '.storybook__tourkit-references-a',
                    mobile: '.storybook__tourkit-references-a',
                },
                meta: {
                    heading: 'Laura 1',
                    descriptions: {
                        desktop: 'Lorem ipsum dolor sit amet.',
                        mobile: 'Lorem ipsum dolor sit amet.',
                    },
                },
            },
            {
                referenceElements: {
                    desktop: '.storybook__tourkit-references-b',
                    mobile: '.storybook__tourkit-references-b',
                },
                meta: {
                    heading: 'Laura 2',
                    descriptions: {
                        desktop: 'Lorem ipsum dolor sit amet.',
                        mobile: 'Lorem ipsum dolor sit amet.',
                    },
                },
            },
            {
                referenceElements: {
                    desktop: '.storybook__tourkit-references-c',
                    mobile: '.storybook__tourkit-references-c',
                },
                meta: {
                    heading: 'Laura 3',
                    descriptions: {
                        desktop: 'Lorem ipsum dolor sit amet.',
                        mobile: 'Lorem ipsum dolor sit amet.',
                    },
                },
            },
            {
                referenceElements: {
                    desktop: '.storybook__tourkit-references-d',
                    mobile: '.storybook__tourkit-references-d',
                },
                meta: {
                    heading: 'Laura 4',
                    descriptions: {
                        desktop: 'Lorem ipsum dolor sit amet.',
                        mobile: 'Lorem ipsum dolor sit amet.',
                    },
                },
            },
        ],
        closeHandler: onClose,
        options: {
            classNames: ['mytour'],
            ...options,
        },
    };
    return _jsx(WpcomTourKit, { config: config });
};
const StoryTour = ({ options = {} }) => {
    const [showTour, setShowTour] = useState(false);
    return (_jsxs("div", { className: "storybook__tourkit", children: [_jsx(References, {}), !showTour && _jsx("button", { onClick: () => setShowTour(true), children: "Start Tour" }), showTour && _jsx(Tour, { onClose: () => setShowTour(false), options: options })] }));
};
export const DefaultWithRating = () => (_jsx(StoryTour, { options: { effects: { arrowIndicator: false }, tourRating: { enabled: true } } }));
export const SpotlightWithRating = () => (_jsx(StoryTour, { options: { effects: { spotlight: {}, arrowIndicator: false }, tourRating: { enabled: true } } }));
