// @automattic/calypso-analytics/dist/cjs/train-tracks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNewRailcarId = exports.recordTrainTracksInteract = exports.recordTrainTracksRender = void 0;
const uuid_1 = require("uuid");
const tracks_1 = require("./tracks");
function recordTrainTracksRender({ railcarId, uiAlgo, uiPosition, fetchAlgo, fetchPosition, query, result, recBlogId, recPostId, recFeedId, recFeedItemId, }) {
    const props = {};
    // Remap and filter undefined props
    Object.entries({
        railcar: railcarId,
        ui_algo: uiAlgo,
        ui_position: uiPosition,
        fetch_algo: fetchAlgo,
        fetch_query: query,
        fetch_position: fetchPosition,
        rec_result: result,
        rec_blog_id: recBlogId,
        rec_post_id: recPostId,
        rec_feed_id: recFeedId,
        rec_feed_item_id: recFeedItemId,
    }).forEach(([key, val]) => val !== undefined && (props[key] = val));
    (0, tracks_1.recordTracksEvent)('calypso_traintracks_render', props);
}
exports.recordTrainTracksRender = recordTrainTracksRender;
function recordTrainTracksInteract({ railcarId, action }) {
    (0, tracks_1.recordTracksEvent)('calypso_traintracks_interact', {
        railcar: railcarId,
        action,
    });
}
exports.recordTrainTracksInteract = recordTrainTracksInteract;
function getNewRailcarId(suffix = 'recommendation') {
    return `${(0, uuid_1.v4)().replace(/-/g, '')}-${suffix}`;
}
exports.getNewRailcarId = getNewRailcarId;
